{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import networkx as nx\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "from collections import Counter\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import hcuppy, json\n",
    "from hcuppy.cci import CCIEngine\n",
    "from hcuppy.ccs import CCSEngine\n",
    "import numpy as np\n",
    "import math\n",
    "import csv\n",
    "from fuzzywuzzy import fuzz \n",
    "from fuzzywuzzy import process \n",
    "import itertools\n",
    "import warnings\n",
    "import gender_guesser.detector as gender\n",
    "from Bio import Entrez\n",
    "from IPython.display import HTML, display\n",
    "from Support.LoadData import (\n",
    "    EFO_parent_mapper,\n",
    "    load_gwas_cat,\n",
    "    load_pubmed_data, \n",
    "    make_timely,\n",
    "    make_clean_CoR,\n",
    "    download_cat\n",
    ")\n",
    "from Support.PubMed import (\n",
    "    build_collective,\n",
    "    build_author,\n",
    "    build_funder,\n",
    "    build_abstract,\n",
    "    build_citation\n",
    ")\n",
    "from Support.Analysis import (\n",
    "    simple_facts,\n",
    "    ancestry_parser,\n",
    "    make_meanfemale_andranks,\n",
    "    make_funders,\n",
    "    mapped_trait_summary\n",
    ")\n",
    "from Support.Figures import (\n",
    "    gwas_growth,\n",
    " #   #choropleth_map,\n",
    "    wordcloud_figure,\n",
    "    plot_heatmap,\n",
    "    plot_bubbles,\n",
    "    boxswarm_plot\n",
    ")\n",
    "from Support.Additional import clean_names\n",
    "from Support.Ancestry import ancestry_cleaner\n",
    "\n",
    "warnings.filterwarnings('ignore')\n",
    "%config InlineBackend.figure_format = 'png'\n",
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download GWAS catalog data from EBI\n",
    "Then, let's dynamically grab the three main curated datasets from the GWAS Catalogue EBI website that we will need for our endeavours ('All Associations','All Studies', 'All Ancestries') and the EFO to Parent trait mapping file from their FTP site:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ebi = 0\n",
    "if ebi == 1:\n",
    "    output_path = os.path.abspath(os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'Catalogue',\n",
    "                                        'Raw'))\n",
    "    ebi_download = 'https://www.ebi.ac.uk/gwas/api/search/downloads/'\n",
    "    download_cat(output_path, ebi_download)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Link the GWAS catalog data to PubMed\n",
    "Lets link the PUBMEDID variable to the PUBMED API and get a series of datasets from that using the support functions written in PubMed.py. Note: collective corresponds mostly consortia and study groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "pubmed = 0\n",
    "if pubmed == 1: \n",
    "    Cat_Studies, Cat_Ancestry, Cat_Ancestry_groupedbyN, Cat_Full = load_gwas_cat()\n",
    "    id_list = set(Cat_Studies['PUBMEDID'].astype(str).tolist())\n",
    "    Entrez.email = 'your@email.com'\n",
    "    papers = Entrez.read(Entrez.efetch(db='pubmed',retmode='xml', id=','.join(id_list)))\n",
    "    build_collective(papers)\n",
    "    build_author(papers)\n",
    "    build_funder(papers)\n",
    "    build_abstract(papers)\n",
    "    #build_citation(id_list,Entrez.email)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading the Data into Memory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the data\n",
    "Cat_Studies, Cat_Ancestry, Cat_Ancestry_groupedbyN, Cat_Full = load_gwas_cat()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The\"load_gwas_cat()\" function loads four datasets:\n",
    "* **\"Cat_Studies\"**: each observation is a studied disease/trait, therefore, one publication may appear multiple times if it has involved multiple disease/trait studies.<br> **INFO:** the citation format, sample (inital, replication), disease/trait, study id. <br>\n",
    "Index(['DATE ADDED TO CATALOG', 'PUBMEDID', 'FIRST AUTHOR', 'DATE', 'JOURNAL',\n",
    "       'LINK', 'STUDY', 'DISEASE/TRAIT', 'INITIAL SAMPLE SIZE',\n",
    "       'REPLICATION SAMPLE SIZE', 'PLATFORM [SNPS PASSING QC]',\n",
    "       'ASSOCIATION COUNT', 'MAPPED_TRAIT', 'MAPPED_TRAIT_URI',\n",
    "       'STUDY ACCESSION', 'GENOTYPING TECHNOLOGY'],\n",
    "      dtype='object')\n",
    "      \n",
    "      \n",
    "* **\"Cat_Ancestry\"**: each observation is a study-ancenstry pair. if a study incorperates multiple ancestries, then it will appear more than once so as to record all the recruited ancestries. <br> **INFO:** study id, detailed sample info(initial, replication, ancestry, size, country of origin, country of recruitment). <br>\n",
    "Index(['STUDY ACCESSION', 'PUBMEDID', 'FIRST AUTHOR', 'DATE',\n",
    "       'INITIAL SAMPLE DESCRIPTION', 'REPLICATION SAMPLE DESCRIPTION', 'STAGE',\n",
    "       'N', 'BROAD ANCESTRAL', 'COUNTRY OF ORIGIN', 'COUNTRY OF RECRUITMENT',\n",
    "       'ADDITONAL ANCESTRY DESCRIPTION', 'Broader', 'Dates'],\n",
    "      dtype='object') \n",
    "      \n",
    "      \n",
    "* **\"Cat_Full\"**: each observation is a SNP-trait pair. if a study focuses on one disease/trait, it finds more than one SNP is associated with that disease/trait, then it should record each SNP in this file for that diseaes/trait in that study. <br>\n",
    "**INFO:** citation format, diseae/trait, sample info, SNP info. <br>\n",
    "Index(['DATE ADDED TO CATALOG', 'PUBMEDID', 'FIRST AUTHOR', 'DATE', 'JOURNAL',\n",
    "       'LINK', 'STUDY', 'DISEASE/TRAIT', 'INITIAL SAMPLE SIZE',\n",
    "       'REPLICATION SAMPLE SIZE', 'REGION', 'CHR_ID', 'CHR_POS',\n",
    "       'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID',\n",
    "       'DOWNSTREAM_GENE_ID', 'SNP_GENE_IDS', 'UPSTREAM_GENE_DISTANCE',\n",
    "       'DOWNSTREAM_GENE_DISTANCE', 'STRONGEST SNP-RISK ALLELE', 'SNPS',\n",
    "       'MERGED', 'SNP_ID_CURRENT', 'CONTEXT', 'INTERGENIC',\n",
    "       'RISK ALLELE FREQUENCY', 'P-VALUE', 'PVALUE_MLOG', 'P-VALUE (TEXT)',\n",
    "       'OR or BETA', '95% CI (TEXT)', 'PLATFORM [SNPS PASSING QC]', 'CNV'],\n",
    "      dtype='object')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ancestry\n",
    "\n",
    "<font color = \"blue\"> Mainly using the \"Cat_Ancestry\" data set\" </font>\n",
    "\n",
    "We use the 'Broad Ancestral Category' field and aggregate from 135 combinations of 17 ancestries to 7 'broader ancestral categories' to calculate a comparable measure to Popejoy and Fullerton.\n",
    "<br>\n",
    "<br>\n",
    "By way of a very simple example: “19,546 British ancestry individuals from 6863 families.” will get cleaned to two seperate fields: “19,546” and “British” which can then be used for further downstream analysis. A slightly more complex example: “2,798 European ancestry individuals, 228 French Canadian founder population individuals” will correspond to two entries of 2798 and 228 in the new ‘Cleaned N’ type variable, corresponding to ‘European’ and ‘French Canadian’ in the ‘Cleaned Ancestry’ type variable respectively"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classify the detailed ancestry in each study into 7 broad ancestry categories.\n",
    "Check that our 'Broader' dictionary is up to date?\n",
    "<font color = 'blue'> In the above, we merge the \"Cat_Ancestry\" file with the \"dict_replacer_broad\" to classify the detailed ancestry into several broad categries of races. Lets first now do a quick check that our 'Broader' dictionary is up to date: </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "      <th>Dates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>7814</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>Klein RJ</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>146</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-03-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7808</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>744</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7809</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>141</td>\n",
       "      <td>Other</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other/Mixed</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7810</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>1</td>\n",
       "      <td>Asian unspecified</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Asian</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7811</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>replication</td>\n",
       "      <td>541</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     STUDY ACCESSION  PUBMEDID   FIRST AUTHOR        DATE  \\\n",
       "7814      GCST000001  15761122       Klein RJ  2005-03-10   \n",
       "7808      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "7809      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "7810      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "7811      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "\n",
       "                             INITIAL SAMPLE DESCRIPTION  \\\n",
       "7814  96 European ancestry cases, 50 European ancest...   \n",
       "7808  381 European ancestry cases, 363 European ance...   \n",
       "7809  381 European ancestry cases, 363 European ance...   \n",
       "7810  381 European ancestry cases, 363 European ance...   \n",
       "7811  381 European ancestry cases, 363 European ance...   \n",
       "\n",
       "                         REPLICATION SAMPLE DESCRIPTION        STAGE    N  \\\n",
       "7814                                                NaN      initial  146   \n",
       "7808  269 European ancestry cases, 272 European ance...      initial  744   \n",
       "7809  269 European ancestry cases, 272 European ance...      initial  141   \n",
       "7810  269 European ancestry cases, 272 European ance...      initial    1   \n",
       "7811  269 European ancestry cases, 272 European ance...  replication  541   \n",
       "\n",
       "        BROAD ANCESTRAL COUNTRY OF ORIGIN COUNTRY OF RECRUITMENT  \\\n",
       "7814           European                NR                     NR   \n",
       "7808           European                NR                   U.S.   \n",
       "7809              Other                NR                   U.S.   \n",
       "7810  Asian unspecified                NR                   U.S.   \n",
       "7811           European                NR                   U.S.   \n",
       "\n",
       "     ADDITONAL ANCESTRY DESCRIPTION      Broader      Dates  \n",
       "7814                            NaN     European 2005-03-10  \n",
       "7808                            NaN     European 2005-09-09  \n",
       "7809                            NaN  Other/Mixed 2005-09-09  \n",
       "7810                            NaN        Asian 2005-09-09  \n",
       "7811                            NaN     European 2005-09-09  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the ancestry data set:\n",
    "Cat_Ancestry.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nice! Looks like our Broad to Broader dictionary is up to date!\n"
     ]
    }
   ],
   "source": [
    "if len(Cat_Ancestry[Cat_Ancestry['Broader'].isnull()]) > 0:\n",
    "    print('Perhaps we need to update the dictionary for new terms? Something like these ones:\\n' +\n",
    "         '\\n'.join(Cat_Ancestry[Cat_Ancestry['Broader'].isnull()]['BROAD ANCESTRAL'].unique()))\n",
    "else:\n",
    "    print('Nice! Looks like our Broad to Broader dictionary is up to date!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aggregated Pattern: Check the sample in the initial and replication stages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## The Initial Stage:\n",
    "Broad_Ancestral_Full_initial = pd.DataFrame(\n",
    "    Cat_Ancestry[Cat_Ancestry.STAGE ==\n",
    "                 'initial'].groupby(['Broader'])['N'].sum())\n",
    "Broad_Ancestral_Full_initial.rename(\n",
    "    columns={'N': 'N (Initial)'}, inplace=True)\n",
    "\n",
    "#Broad_Ancestral_Full_initial.head(10)\n",
    "# 7 broad race categories\n",
    "\n",
    "\n",
    "## The Replication Stage:\n",
    "Broad_Ancestral_Full_replication = pd.DataFrame(\n",
    "    Cat_Ancestry[Cat_Ancestry.STAGE ==\n",
    "                 'replication'].groupby(['Broader'])['N'].sum())\n",
    "Broad_Ancestral_Full_replication.rename(\n",
    "    columns={'N': 'N (Replication)'}, inplace=True)\n",
    "\n",
    "#Broad_Ancestral_Full_replication.head(10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color = 'blue' size = 4pt> Merge the initial and replication sample into a total one. </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N (Initial)</th>\n",
       "      <th>N (Replication)</th>\n",
       "      <th>Total</th>\n",
       "      <th>% Discovery</th>\n",
       "      <th>% Replication</th>\n",
       "      <th>% Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Broader</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>African</td>\n",
       "      <td>1428501</td>\n",
       "      <td>387754</td>\n",
       "      <td>1816255</td>\n",
       "      <td>0.484830</td>\n",
       "      <td>0.422394</td>\n",
       "      <td>0.469998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>African Am./Caribbean</td>\n",
       "      <td>4002921</td>\n",
       "      <td>1207839</td>\n",
       "      <td>5210760</td>\n",
       "      <td>1.358583</td>\n",
       "      <td>1.315742</td>\n",
       "      <td>1.348406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Asian</td>\n",
       "      <td>15314535</td>\n",
       "      <td>12634906</td>\n",
       "      <td>27949441</td>\n",
       "      <td>5.197720</td>\n",
       "      <td>13.763655</td>\n",
       "      <td>7.232571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>European</td>\n",
       "      <td>249236131</td>\n",
       "      <td>63907946</td>\n",
       "      <td>313144077</td>\n",
       "      <td>84.590208</td>\n",
       "      <td>69.617213</td>\n",
       "      <td>81.033350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Hispanic/Latin American</td>\n",
       "      <td>3075425</td>\n",
       "      <td>1203640</td>\n",
       "      <td>4279065</td>\n",
       "      <td>1.043793</td>\n",
       "      <td>1.311168</td>\n",
       "      <td>1.107308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>In Part Not Recorded</td>\n",
       "      <td>19211756</td>\n",
       "      <td>11513163</td>\n",
       "      <td>30724919</td>\n",
       "      <td>6.520429</td>\n",
       "      <td>12.541701</td>\n",
       "      <td>7.950791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Other/Mixed</td>\n",
       "      <td>2370189</td>\n",
       "      <td>943810</td>\n",
       "      <td>3313999</td>\n",
       "      <td>0.804437</td>\n",
       "      <td>1.028126</td>\n",
       "      <td>0.857575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         N (Initial)  N (Replication)      Total  % Discovery  \\\n",
       "Broader                                                                         \n",
       "African                      1428501           387754    1816255     0.484830   \n",
       "African Am./Caribbean        4002921          1207839    5210760     1.358583   \n",
       "Asian                       15314535         12634906   27949441     5.197720   \n",
       "European                   249236131         63907946  313144077    84.590208   \n",
       "Hispanic/Latin American      3075425          1203640    4279065     1.043793   \n",
       "In Part Not Recorded        19211756         11513163   30724919     6.520429   \n",
       "Other/Mixed                  2370189           943810    3313999     0.804437   \n",
       "\n",
       "                         % Replication    % Total  \n",
       "Broader                                            \n",
       "African                       0.422394   0.469998  \n",
       "African Am./Caribbean         1.315742   1.348406  \n",
       "Asian                        13.763655   7.232571  \n",
       "European                     69.617213  81.033350  \n",
       "Hispanic/Latin American       1.311168   1.107308  \n",
       "In Part Not Recorded         12.541701   7.950791  \n",
       "Other/Mixed                   1.028126   0.857575  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Broad_Ancestral_Full = pd.merge(Broad_Ancestral_Full_initial,\n",
    "                                Broad_Ancestral_Full_replication,\n",
    "                                left_index=True, right_index=True)\n",
    "Broad_Ancestral_Full['Total'] = Broad_Ancestral_Full['N (Initial)'] + \\\n",
    "    Broad_Ancestral_Full['N (Replication)']\n",
    "Broad_Ancestral_Full['% Discovery'] = (\n",
    "    Broad_Ancestral_Full['N (Initial)'] /\n",
    "    Broad_Ancestral_Full['N (Initial)'].sum()) * 100\n",
    "Broad_Ancestral_Full['% Replication'] = (\n",
    "    Broad_Ancestral_Full['N (Replication)'] /\n",
    "    Broad_Ancestral_Full['N (Replication)'].sum()) * 100\n",
    "Broad_Ancestral_Full['% Total'] = (Broad_Ancestral_Full['Total'] /\n",
    "                                   Broad_Ancestral_Full['Total'].sum()) * 100\n",
    "Broad_Ancestral_Full.to_csv(os.path.abspath(\n",
    "                            os.path.join('__file__',\n",
    "                                         '../..',\n",
    "                                         'tables',\n",
    "                                         'Broad_Ancestral_Full.csv')))\n",
    "Broad_Ancestral_Full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color = 'blue' size = 4pt> Drop the 'in part not recorded' rows (i.e. where the Broad Ancestral Category contains at least one NR):\n",
    "    </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N (Initial)</th>\n",
       "      <th>N (Replication)</th>\n",
       "      <th>Total</th>\n",
       "      <th>% Discovery</th>\n",
       "      <th>% Replication</th>\n",
       "      <th>% Total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Broader</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>African</td>\n",
       "      <td>1428501</td>\n",
       "      <td>387754</td>\n",
       "      <td>1816255</td>\n",
       "      <td>0.518648</td>\n",
       "      <td>0.482967</td>\n",
       "      <td>0.510595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>African Am./Caribbean</td>\n",
       "      <td>4002921</td>\n",
       "      <td>1207839</td>\n",
       "      <td>5210760</td>\n",
       "      <td>1.453347</td>\n",
       "      <td>1.504422</td>\n",
       "      <td>1.464875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Asian</td>\n",
       "      <td>15314535</td>\n",
       "      <td>12634906</td>\n",
       "      <td>27949441</td>\n",
       "      <td>5.560274</td>\n",
       "      <td>15.737392</td>\n",
       "      <td>7.857288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>European</td>\n",
       "      <td>249236131</td>\n",
       "      <td>63907946</td>\n",
       "      <td>313144077</td>\n",
       "      <td>90.490582</td>\n",
       "      <td>79.600465</td>\n",
       "      <td>88.032642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Hispanic/Latin American</td>\n",
       "      <td>3075425</td>\n",
       "      <td>1203640</td>\n",
       "      <td>4279065</td>\n",
       "      <td>1.116600</td>\n",
       "      <td>1.499192</td>\n",
       "      <td>1.202952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Other/Mixed</td>\n",
       "      <td>2370189</td>\n",
       "      <td>943810</td>\n",
       "      <td>3313999</td>\n",
       "      <td>0.860549</td>\n",
       "      <td>1.175561</td>\n",
       "      <td>0.931648</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         N (Initial)  N (Replication)      Total  % Discovery  \\\n",
       "Broader                                                                         \n",
       "African                      1428501           387754    1816255     0.518648   \n",
       "African Am./Caribbean        4002921          1207839    5210760     1.453347   \n",
       "Asian                       15314535         12634906   27949441     5.560274   \n",
       "European                   249236131         63907946  313144077    90.490582   \n",
       "Hispanic/Latin American      3075425          1203640    4279065     1.116600   \n",
       "Other/Mixed                  2370189           943810    3313999     0.860549   \n",
       "\n",
       "                         % Replication    % Total  \n",
       "Broader                                            \n",
       "African                       0.482967   0.510595  \n",
       "African Am./Caribbean         1.504422   1.464875  \n",
       "Asian                        15.737392   7.857288  \n",
       "European                     79.600465  88.032642  \n",
       "Hispanic/Latin American       1.499192   1.202952  \n",
       "Other/Mixed                   1.175561   0.931648  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Broad_Ancestral_NoNR = Broad_Ancestral_Full[['N (Initial)',\n",
    "                                             'N (Replication)',\n",
    "                                             'Total']]\n",
    "Broad_Ancestral_NoNR = Broad_Ancestral_NoNR.drop('In Part Not Recorded', axis = 0)\n",
    "Broad_Ancestral_NoNR['Total'] = Broad_Ancestral_NoNR['N (Initial)'] + \\\n",
    "    Broad_Ancestral_NoNR['N (Replication)']\n",
    "Broad_Ancestral_NoNR['% Discovery'] = (\n",
    "    Broad_Ancestral_NoNR['N (Initial)'] /\n",
    "    Broad_Ancestral_NoNR['N (Initial)'].sum()) * 100\n",
    "Broad_Ancestral_NoNR['% Replication'] = (\n",
    "    Broad_Ancestral_NoNR['N (Replication)'] /\n",
    "    Broad_Ancestral_NoNR['N (Replication)'].sum()) * 100\n",
    "Broad_Ancestral_NoNR['% Total'] = (Broad_Ancestral_NoNR['Total'] /\n",
    "                                   Broad_Ancestral_NoNR['Total'].sum()) * 100\n",
    "Broad_Ancestral_NoNR.to_csv(os.path.abspath(\n",
    "                            os.path.join('__file__', '../..', 'tables',\n",
    "                                         'Broad_Ancestral_NoNR.csv')))\n",
    "Broad_Ancestral_NoNR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color = 'blue' size = 4pt> Time Trend  </font> <br>\n",
    "We can also analyze how these aggregates change over time, and this was a key feature of [Popejoy and Fullerton (2016)](https://www.nature.com/news/genomics-is-failing-on-diversity-1.20759). We can provide a much more granular arguement (merely remove ```'_NoNR'``` from the cell below to undertake an equivlent analysis with rows which contain some in part NR ancestries):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "</style><table id=\"T_ef304734_0001_11ea_8a7a_004e019a962b\" ><thead>    <tr>        <th class=\"blank level0\" ></th>        <th class=\"col_heading level0 col0\" >European</th>        <th class=\"col_heading level0 col1\" >Asian</th>        <th class=\"col_heading level0 col2\" >African</th>        <th class=\"col_heading level0 col3\" >Hispanic/Latin American</th>        <th class=\"col_heading level0 col4\" >Other/Mixed</th>        <th class=\"col_heading level0 col5\" >African Am./Caribbean</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row0\" class=\"row_heading level0 row0\" >2007</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow0_col0\" class=\"data row0 col0\" >95.4688</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow0_col1\" class=\"data row0 col1\" >2.13984</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow0_col2\" class=\"data row0 col2\" >0.00542792</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow0_col3\" class=\"data row0 col3\" >0.715245</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow0_col4\" class=\"data row0 col4\" >1.18391</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow0_col5\" class=\"data row0 col5\" >0.486807</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row1\" class=\"row_heading level0 row1\" >2008</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow1_col0\" class=\"data row1 col0\" >95.2884</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow1_col1\" class=\"data row1 col1\" >2.94597</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow1_col2\" class=\"data row1 col2\" >0</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow1_col3\" class=\"data row1 col3\" >0.00180063</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow1_col4\" class=\"data row1 col4\" >1.21894</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow1_col5\" class=\"data row1 col5\" >0.544872</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row2\" class=\"row_heading level0 row2\" >2009</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow2_col0\" class=\"data row2 col0\" >88.4027</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow2_col1\" class=\"data row2 col1\" >6.96558</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow2_col2\" class=\"data row2 col2\" >0.253076</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow2_col3\" class=\"data row2 col3\" >0.219723</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow2_col4\" class=\"data row2 col4\" >3.29458</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow2_col5\" class=\"data row2 col5\" >0.86432</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row3\" class=\"row_heading level0 row3\" >2010</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow3_col0\" class=\"data row3 col0\" >86.7933</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow3_col1\" class=\"data row3 col1\" >9.91171</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow3_col2\" class=\"data row3 col2\" >0.267259</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow3_col3\" class=\"data row3 col3\" >0.0583916</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow3_col4\" class=\"data row3 col4\" >2.47723</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow3_col5\" class=\"data row3 col5\" >0.49206</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row4\" class=\"row_heading level0 row4\" >2011</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow4_col0\" class=\"data row4 col0\" >78.265</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow4_col1\" class=\"data row4 col1\" >15.7991</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow4_col2\" class=\"data row4 col2\" >0.15655</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow4_col3\" class=\"data row4 col3\" >0.396955</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow4_col4\" class=\"data row4 col4\" >1.70627</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow4_col5\" class=\"data row4 col5\" >3.6762</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row5\" class=\"row_heading level0 row5\" >2012</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow5_col0\" class=\"data row5 col0\" >71.8522</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow5_col1\" class=\"data row5 col1\" >19.0443</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow5_col2\" class=\"data row5 col2\" >0.306187</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow5_col3\" class=\"data row5 col3\" >0.86527</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow5_col4\" class=\"data row5 col4\" >3.30234</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow5_col5\" class=\"data row5 col5\" >4.62965</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row6\" class=\"row_heading level0 row6\" >2013</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow6_col0\" class=\"data row6 col0\" >82.3989</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow6_col1\" class=\"data row6 col1\" >11.6072</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow6_col2\" class=\"data row6 col2\" >0.385825</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow6_col3\" class=\"data row6 col3\" >0.770707</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow6_col4\" class=\"data row6 col4\" >0.61389</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow6_col5\" class=\"data row6 col5\" >4.22351</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row7\" class=\"row_heading level0 row7\" >2014</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow7_col0\" class=\"data row7 col0\" >76.6765</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow7_col1\" class=\"data row7 col1\" >18.0765</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow7_col2\" class=\"data row7 col2\" >0.851494</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow7_col3\" class=\"data row7 col3\" >1.11809</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow7_col4\" class=\"data row7 col4\" >0.949118</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow7_col5\" class=\"data row7 col5\" >2.32834</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row8\" class=\"row_heading level0 row8\" >2015</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow8_col0\" class=\"data row8 col0\" >88.0835</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow8_col1\" class=\"data row8 col1\" >9.04736</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow8_col2\" class=\"data row8 col2\" >0.26423</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow8_col3\" class=\"data row8 col3\" >0.700762</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow8_col4\" class=\"data row8 col4\" >0.49246</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow8_col5\" class=\"data row8 col5\" >1.41172</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row9\" class=\"row_heading level0 row9\" >2016</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow9_col0\" class=\"data row9 col0\" >90.8446</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow9_col1\" class=\"data row9 col1\" >4.47222</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow9_col2\" class=\"data row9 col2\" >0.171732</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow9_col3\" class=\"data row9 col3\" >1.58114</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow9_col4\" class=\"data row9 col4\" >1.01288</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow9_col5\" class=\"data row9 col5\" >1.91738</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row10\" class=\"row_heading level0 row10\" >2017</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow10_col0\" class=\"data row10 col0\" >89.6856</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow10_col1\" class=\"data row10 col1\" >5.35262</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow10_col2\" class=\"data row10 col2\" >0.503509</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow10_col3\" class=\"data row10 col3\" >1.95875</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow10_col4\" class=\"data row10 col4\" >0.540159</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow10_col5\" class=\"data row10 col5\" >1.95937</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_ef304734_0001_11ea_8a7a_004e019a962blevel0_row11\" class=\"row_heading level0 row11\" >2018</th>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow11_col0\" class=\"data row11 col0\" >86.9283</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow11_col1\" class=\"data row11 col1\" >10.4085</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow11_col2\" class=\"data row11 col2\" >0.464814</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow11_col3\" class=\"data row11 col3\" >0.793711</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow11_col4\" class=\"data row11 col4\" >0.531062</td>\n",
       "                        <td id=\"T_ef304734_0001_11ea_8a7a_004e019a962brow11_col5\" class=\"data row11 col5\" >0.873615</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x14ffd91e188>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index = [x for x in range(2007, 2019)]\n",
    "columns = ['European', 'Asian', 'African', 'Hispanic/Latin American',\n",
    "           'Other/Mixed', 'African Am./Caribbean']\n",
    "Cat_Ancestry_NoNR = Cat_Ancestry[\n",
    "    Cat_Ancestry['Broader'] != 'In Part Not Recorded']\n",
    "Broad_Ancestral_Time_NoNR_PC = pd.DataFrame(index=index, columns=columns)\n",
    "for year in range(2007, 2019):\n",
    "    for broaderancestry in Broad_Ancestral_NoNR.index.tolist():\n",
    "        Broad_Ancestral_Time_NoNR_PC[broaderancestry.strip()][year] =\\\n",
    "            (Cat_Ancestry_NoNR[(\n",
    "                Cat_Ancestry_NoNR['DATE'].str.contains(str(year))) &\n",
    "                (Cat_Ancestry_NoNR['Broader'] ==\n",
    "                 broaderancestry)]['N'].sum() /\n",
    "             Cat_Ancestry_NoNR[\n",
    "             (Cat_Ancestry_NoNR['DATE'].str.contains(\n",
    "                 str(year)))]['N'].sum()) * 100\n",
    "Broad_Ancestral_Time_NoNR_PC.to_csv(os.path.abspath(\n",
    "                            os.path.join('__file__', '../..', 'tables',\n",
    "                                         'Broad_Ancestral_Time_NoNR_PC.csv')))\n",
    "Broad_Ancestral_Time_NoNR_PC.style"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ancestry for Each Study \n",
    "\n",
    "We expect to construct a data set with the following format: <br>\n",
    "**basic info:** <br>\n",
    "- Study PubMedID year EFO\n",
    "\n",
    "**ancestry by race and stages:** <br>\n",
    "- initial: (European Asian, African Hispanic African American Other)\n",
    "- replication: (European Asian, African Hispanic African American Other) \n",
    "\n",
    "**funding per PubmedID:** <br>\n",
    "- Funding in total\n",
    "- Funding in US\n",
    "- Funding in ARRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['STUDY ACCESSION', 'PUBMEDID', 'FIRST AUTHOR', 'DATE',\n",
      "       'INITIAL SAMPLE DESCRIPTION', 'REPLICATION SAMPLE DESCRIPTION', 'STAGE',\n",
      "       'N', 'BROAD ANCESTRAL', 'COUNTRY OF ORIGIN', 'COUNTRY OF RECRUITMENT',\n",
      "       'ADDITONAL ANCESTRY DESCRIPTION', 'Broader', 'Dates'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "      <th>Dates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>7814</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>Klein RJ</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>146</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-03-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7808</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>744</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     STUDY ACCESSION  PUBMEDID   FIRST AUTHOR        DATE  \\\n",
       "7814      GCST000001  15761122       Klein RJ  2005-03-10   \n",
       "7808      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "\n",
       "                             INITIAL SAMPLE DESCRIPTION  \\\n",
       "7814  96 European ancestry cases, 50 European ancest...   \n",
       "7808  381 European ancestry cases, 363 European ance...   \n",
       "\n",
       "                         REPLICATION SAMPLE DESCRIPTION    STAGE    N  \\\n",
       "7814                                                NaN  initial  146   \n",
       "7808  269 European ancestry cases, 272 European ance...  initial  744   \n",
       "\n",
       "     BROAD ANCESTRAL COUNTRY OF ORIGIN COUNTRY OF RECRUITMENT  \\\n",
       "7814        European                NR                     NR   \n",
       "7808        European                NR                   U.S.   \n",
       "\n",
       "     ADDITONAL ANCESTRY DESCRIPTION   Broader      Dates  \n",
       "7814                            NaN  European 2005-03-10  \n",
       "7808                            NaN  European 2005-09-09  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the ancestry data set:\n",
    "print(Cat_Ancestry.columns)\n",
    "Cat_Ancestry.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Is the STUDY-ACCESSION the Unique Identifier in the Acenstry Data? True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>DATE</th>\n",
       "      <th>initial_African</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>replication_African</th>\n",
       "      <th>replication_African Am./Caribbean</th>\n",
       "      <th>replication_Asian</th>\n",
       "      <th>replication_European</th>\n",
       "      <th>replication_Hispanic/Latin American</th>\n",
       "      <th>replication_In Part Not Recorded</th>\n",
       "      <th>replication_Other/Mixed</th>\n",
       "      <th>initial_N</th>\n",
       "      <th>replication_N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>744</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>541</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>886</td>\n",
       "      <td>664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GCST000003</td>\n",
       "      <td>16614226</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>866</td>\n",
       "      <td>0</td>\n",
       "      <td>10527</td>\n",
       "      <td>0</td>\n",
       "      <td>229</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>11622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GCST000004</td>\n",
       "      <td>16648850</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4851</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>4851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GCST000005</td>\n",
       "      <td>17052657</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID        DATE  initial_African  \\\n",
       "0      GCST000001  15761122  2005-03-10                0   \n",
       "1      GCST000002  16252231  2005-09-09                0   \n",
       "2      GCST000003  16614226  2006-04-14                0   \n",
       "3      GCST000004  16648850  2006-04-30                0   \n",
       "4      GCST000005  17052657  2006-09-28                0   \n",
       "\n",
       "   initial_African Am./Caribbean  initial_Asian  initial_European  \\\n",
       "0                              0              0               146   \n",
       "1                              0              1               744   \n",
       "2                              0              0                 0   \n",
       "3                              0              0               200   \n",
       "4                              0              0               537   \n",
       "\n",
       "   initial_Hispanic/Latin American  initial_In Part Not Recorded  \\\n",
       "0                                0                             0   \n",
       "1                                0                             0   \n",
       "2                                0                           694   \n",
       "3                                0                             0   \n",
       "4                                0                             0   \n",
       "\n",
       "   initial_Other/Mixed  replication_African  \\\n",
       "0                    0                    0   \n",
       "1                  141                    0   \n",
       "2                    0                    0   \n",
       "3                    0                    0   \n",
       "4                    0                    0   \n",
       "\n",
       "   replication_African Am./Caribbean  replication_Asian  replication_European  \\\n",
       "0                                  0                  0                     0   \n",
       "1                                  0                  1                   541   \n",
       "2                                866                  0                 10527   \n",
       "3                                  0                  0                  4851   \n",
       "4                                  0                  0                     0   \n",
       "\n",
       "   replication_Hispanic/Latin American  replication_In Part Not Recorded  \\\n",
       "0                                    0                                 0   \n",
       "1                                    0                                 0   \n",
       "2                                    0                               229   \n",
       "3                                    0                                 0   \n",
       "4                                    0                                 0   \n",
       "\n",
       "   replication_Other/Mixed  initial_N  replication_N  \n",
       "0                        0        146              0  \n",
       "1                      122        886            664  \n",
       "2                        0        694          11622  \n",
       "3                        0        200           4851  \n",
       "4                        0        537              0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry = Cat_Ancestry[['STUDY ACCESSION', 'PUBMEDID', 'DATE', 'INITIAL SAMPLE DESCRIPTION', 'REPLICATION SAMPLE DESCRIPTION',\n",
    "                         'STAGE', 'N', 'Broader']]\n",
    "\n",
    "Ancestry = Ancestry.groupby(['STUDY ACCESSION', 'STAGE', 'DATE', 'Broader', 'PUBMEDID'])['N'].sum().reset_index()\n",
    "\n",
    "Ancestry = Ancestry.pivot_table(\n",
    "         index = ['STUDY ACCESSION', 'PUBMEDID', 'Broader', 'DATE'],\n",
    "         columns = 'STAGE',\n",
    "         values = 'N', \n",
    "         fill_value = 0)\n",
    "\n",
    "Ancestry = Ancestry.unstack(-2, fill_value= 0)\n",
    "Ancestry.columns = Ancestry.columns.map('_'.join).str.strip('|')\n",
    "\n",
    "Ancestry['initial_N'] = Ancestry.iloc[:, Ancestry.columns.str.startswith('initial_')].sum(axis = 1)\n",
    "Ancestry['replication_N'] = Ancestry.iloc[:, Ancestry.columns.str.startswith('replication_')].sum(axis = 1)\n",
    "\n",
    "Ancestry.reset_index(inplace = True)\n",
    "\n",
    "print('Is the STUDY-ACCESSION the Unique Identifier in the Acenstry Data?', ~Ancestry.duplicated('STUDY ACCESSION').any())\n",
    "Ancestry.head()       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge Funding info data for each article."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>Agency</th>\n",
       "      <th>GrantCountry</th>\n",
       "      <th>GrantID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>24029427</td>\n",
       "      <td>NIDDK NIH HHS</td>\n",
       "      <td>United States</td>\n",
       "      <td>U01DK057303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>24029427</td>\n",
       "      <td>NIDDK NIH HHS</td>\n",
       "      <td>United States</td>\n",
       "      <td>U01 DK057304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>24029427</td>\n",
       "      <td>Medical Research Council</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>MR/K003364/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>24029427</td>\n",
       "      <td>Wellcome Trust</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>WT090355/A/09/Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>24029427</td>\n",
       "      <td>NCRR NIH HHS</td>\n",
       "      <td>United States</td>\n",
       "      <td>M01-RR-01346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PUBMEDID                    Agency    GrantCountry          GrantID\n",
       "0  24029427             NIDDK NIH HHS   United States      U01DK057303\n",
       "1  24029427             NIDDK NIH HHS   United States     U01 DK057304\n",
       "2  24029427  Medical Research Council  United Kingdom     MR/K003364/1\n",
       "3  24029427            Wellcome Trust  United Kingdom  WT090355/A/09/Z\n",
       "4  24029427              NCRR NIH HHS   United States     M01-RR-01346"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Funding = pd.read_csv(os.path.abspath(\n",
    "                           os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'PUBMED',\n",
    "                                        'Pubmed_FunderInfo.csv')),\n",
    "                           header=0, sep=',', encoding='utf-8',\n",
    "                           index_col=False)\n",
    "Funding.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Is PubMedId the unique identifier in the funding data? True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>Funded_US</th>\n",
       "      <th>Funded</th>\n",
       "      <th>Funded_NIH</th>\n",
       "      <th>Funded_AHRQ</th>\n",
       "      <th>Funded_HHS</th>\n",
       "      <th>Funded_ARRA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>15761122</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>16252231</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>16614226</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>16648850</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>17052657</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PUBMEDID  Funded_US  Funded  Funded_NIH  Funded_AHRQ  Funded_HHS  \\\n",
       "0  15761122          6       6           6            0           6   \n",
       "1  16252231          8       8           8            0           8   \n",
       "2  16614226         11      11          11            0          11   \n",
       "3  16648850          3       3           3            0           3   \n",
       "4  17052657          1       2           1            0           1   \n",
       "\n",
       "   Funded_ARRA  \n",
       "0            6  \n",
       "1            8  \n",
       "2           11  \n",
       "3            3  \n",
       "4            1  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Funding['Funded_US'] = (Funding['GrantCountry'] == 'United States').astype(int)\n",
    "Funding['Funded'] = (Funding['GrantCountry'] != '').astype(int)\n",
    "Funding['Funded_NIH'] = (Funding['Agency'].str.contains('(NIH)'))\n",
    "Funding['Funded_AHRQ'] = (Funding['Agency'].str.contains('(AHRQ)'))\n",
    "Funding['Funded_HHS'] = (Funding['Agency'].str.contains('(HHS)'))\n",
    "Funding['Funded_ARRA'] = ((Funding['Funded_NIH'] == True) | (Funding['Funded_HHS'] == True) | (Funding['Funded_AHRQ'] == True)).astype(int)\n",
    "\n",
    "Funding.loc[Funding[\"Funded_NIH\"].apply(math.isnan), 'Funded_NIH'] = 0 \n",
    "Funding.loc[Funding[\"Funded_HHS\"].apply(math.isnan), 'Funded_HHS'] = 0 \n",
    "Funding.loc[Funding[\"Funded_AHRQ\"].apply(math.isnan), 'Funded_AHRQ'] = 0 \n",
    "\n",
    "Funding.replace([True, False], [1,0], inplace = True)\n",
    "Funding = Funding.groupby('PUBMEDID').sum().reset_index()\n",
    "print('Is PubMedId the unique identifier in the funding data?', ~Funding.duplicated('PUBMEDID').any())\n",
    "Funding.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>DATE</th>\n",
       "      <th>initial_African</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>...</th>\n",
       "      <th>replication_Other/Mixed</th>\n",
       "      <th>initial_N</th>\n",
       "      <th>replication_N</th>\n",
       "      <th>Funded_US</th>\n",
       "      <th>Funded</th>\n",
       "      <th>Funded_NIH</th>\n",
       "      <th>Funded_AHRQ</th>\n",
       "      <th>Funded_HHS</th>\n",
       "      <th>Funded_ARRA</th>\n",
       "      <th>Funding_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>funding_info_y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>744</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>...</td>\n",
       "      <td>122</td>\n",
       "      <td>886</td>\n",
       "      <td>664</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>funding_info_y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GCST000003</td>\n",
       "      <td>16614226</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>11622</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>funding_info_y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GCST000004</td>\n",
       "      <td>16648850</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>4851</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>funding_info_y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GCST000005</td>\n",
       "      <td>17052657</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>funding_info_y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID        DATE  initial_African  \\\n",
       "0      GCST000001  15761122  2005-03-10                0   \n",
       "1      GCST000002  16252231  2005-09-09                0   \n",
       "2      GCST000003  16614226  2006-04-14                0   \n",
       "3      GCST000004  16648850  2006-04-30                0   \n",
       "4      GCST000005  17052657  2006-09-28                0   \n",
       "\n",
       "   initial_African Am./Caribbean  initial_Asian  initial_European  \\\n",
       "0                              0              0               146   \n",
       "1                              0              1               744   \n",
       "2                              0              0                 0   \n",
       "3                              0              0               200   \n",
       "4                              0              0               537   \n",
       "\n",
       "   initial_Hispanic/Latin American  initial_In Part Not Recorded  \\\n",
       "0                                0                             0   \n",
       "1                                0                             0   \n",
       "2                                0                           694   \n",
       "3                                0                             0   \n",
       "4                                0                             0   \n",
       "\n",
       "   initial_Other/Mixed  ...  replication_Other/Mixed  initial_N  \\\n",
       "0                    0  ...                        0        146   \n",
       "1                  141  ...                      122        886   \n",
       "2                    0  ...                        0        694   \n",
       "3                    0  ...                        0        200   \n",
       "4                    0  ...                        0        537   \n",
       "\n",
       "   replication_N  Funded_US  Funded  Funded_NIH  Funded_AHRQ  Funded_HHS  \\\n",
       "0              0        6.0     6.0         6.0          0.0         6.0   \n",
       "1            664        8.0     8.0         8.0          0.0         8.0   \n",
       "2          11622       11.0    11.0        11.0          0.0        11.0   \n",
       "3           4851        3.0     3.0         3.0          0.0         3.0   \n",
       "4              0        1.0     2.0         1.0          0.0         1.0   \n",
       "\n",
       "   Funded_ARRA   Funding_merge  \n",
       "0          6.0  funding_info_y  \n",
       "1          8.0  funding_info_y  \n",
       "2         11.0  funding_info_y  \n",
       "3          3.0  funding_info_y  \n",
       "4          1.0  funding_info_y  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry = pd.merge(Ancestry, Funding, on = 'PUBMEDID',how = 'left',  \n",
    "                               indicator = 'Funding_merge', validate = 'm:1')\n",
    "\n",
    "Ancestry.loc[Ancestry[\"Funded_US\"].apply(math.isnan), 'Funded_US'] = 0 \n",
    "Ancestry.loc[Ancestry[\"Funded\"].apply(math.isnan), 'Funded'] = 0 \n",
    "Ancestry.loc[Ancestry[\"Funded_NIH\"].apply(math.isnan), 'Funded_NIH'] = 0 \n",
    "Ancestry.loc[Ancestry[\"Funded_HHS\"].apply(math.isnan), 'Funded_HHS'] = 0 \n",
    "Ancestry.loc[Ancestry[\"Funded_AHRQ\"].apply(math.isnan), 'Funded_AHRQ'] = 0 \n",
    "Ancestry.loc[Ancestry[\"Funded_ARRA\"].apply(math.isnan), 'Funded_ARRA'] = 0 \n",
    "\n",
    "Ancestry.replace({'Funding_merge': {'both':'funding_info_y', 'left_only': 'funding_info_n'}}, inplace = True)\n",
    "\n",
    "Ancestry.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ancestry (By Disease/Trait)\n",
    "\n",
    "<font color = \"blue\"> mainly using the \"cat_ancstry\", \"EFO-ICD\", and \"ICD-CCS\" data file </font>\n",
    "<br>\n",
    "<br>\n",
    "In this section, we will match each study (*NOT Publication*) to a specific CCS code through the EFO term (disease/trait) explicitly mentioned in that study. Then we are able to calculate the <font color = \"red\"> diversity by disease. </font>\n",
    "<br>\n",
    "The \"EFO-ICD\" dictionary is downloaded from https://www.ebi.ac.uk/spot/oxo/search, saved in \"data/support\" folder\n",
    "* efo2icd10-mappings_dis1.csv\n",
    "* efo2icd10-mappings_dis2.csv\n",
    "* efo2icd10-mappings_dis3.csv <br>\n",
    "<font color = 'blue'>The distance tells you how many times we have seen this mapping and the distance is a how many hops across other mappings you need to go to find this mapping. Distance 1 is a direct mapping, the greater the distance the less likely it is that a mapping holds true </font>\n",
    "<br>\n",
    "\n",
    "The python package \"[hcuppy](https://pypi.org/project/hcuppy/)\" helps to convert the ICD code to the CCS code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The shape of the EFO2ICD10 Data: (7000, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EFO</th>\n",
       "      <th>label</th>\n",
       "      <th>ICD10</th>\n",
       "      <th>mapped_label</th>\n",
       "      <th>mapping_source_prefix</th>\n",
       "      <th>mapping_target_prefix</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>EFO:0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>ICD10CM:M73</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>EFO:0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>ICD10CM:M75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>EFO:0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>ICD10CM:M70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>EFO:0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>ICD10CM:M79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>EFO:0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>ICD10CM:C63.2</td>\n",
       "      <td>Malignant neoplasm of scrotum</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           EFO                label          ICD10  \\\n",
       "0  EFO:0009470  soft tissue disease    ICD10CM:M73   \n",
       "1  EFO:0009470  soft tissue disease    ICD10CM:M75   \n",
       "2  EFO:0009470  soft tissue disease    ICD10CM:M70   \n",
       "3  EFO:0009470  soft tissue disease    ICD10CM:M79   \n",
       "4  EFO:0003865      kidney neoplasm  ICD10CM:C63.2   \n",
       "\n",
       "                    mapped_label mapping_source_prefix mapping_target_prefix  \\\n",
       "0                            NaN                   EFO               ICD10CM   \n",
       "1                            NaN                   EFO               ICD10CM   \n",
       "2                            NaN                   EFO               ICD10CM   \n",
       "3                            NaN                   EFO               ICD10CM   \n",
       "4  Malignant neoplasm of scrotum             EFO,MONDO               ICD10CM   \n",
       "\n",
       "   distance  \n",
       "0         1  \n",
       "1         1  \n",
       "2         1  \n",
       "3         1  \n",
       "4         3  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFO2ICD10 = pd.read_csv(os.path.abspath(\n",
    "                           os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'Support',\n",
    "                                        'efo2icd10-mappings_dis3.csv')),\n",
    "                           header=0, sep=',', encoding='utf-8',\n",
    "                           index_col=False)\n",
    "#EFO2ICD10.head()\n",
    "\n",
    "# rename the column title\n",
    "EFO2ICD10.rename(columns = {'curie_id': 'EFO', 'mapped_curie': 'ICD10'}, inplace = True)\n",
    "\n",
    "print(\"The shape of the EFO2ICD10 Data:\", EFO2ICD10.shape)\n",
    "\n",
    "EFO2ICD10.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To assign each study a corresponding CCS code, we need to translate the EFO term in each study to ICD10, and then from ICD10 to CCS code.\n",
    "\n",
    "(1) Create a EFO to CCS dictionary (via ICD10) \n",
    "* We first extract the <font color = \"blue\"> EFO ID </font> from the <font color = \"blue\"> EFO </font> column in the **EFO2ICD10** file. The pattern is that: \"EFO:0009470\" --> '0009470'.\n",
    "* Then we match the above **EFO2ICD10** file to the <font color = \"blue\"> CCS </font> code using the <font color = \"blue\"> hcuppy </font> package.\n",
    "\n",
    "(2) match the \"CAT_Ancestry\" file to the EFO-CCS dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EFO</th>\n",
       "      <th>label</th>\n",
       "      <th>ICD10</th>\n",
       "      <th>mapped_label</th>\n",
       "      <th>mapping_source_prefix</th>\n",
       "      <th>mapping_target_prefix</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M73</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C63.2</td>\n",
       "      <td>Malignant neoplasm of scrotum</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C58</td>\n",
       "      <td>Malignant neoplasm of placenta</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C73</td>\n",
       "      <td>Malignant neoplasm of thyroid gland</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>D23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       EFO                label  ICD10                         mapped_label  \\\n",
       "0  0009470  soft tissue disease    M73                                  NaN   \n",
       "1  0009470  soft tissue disease    M75                                  NaN   \n",
       "2  0009470  soft tissue disease    M70                                  NaN   \n",
       "3  0009470  soft tissue disease    M79                                  NaN   \n",
       "4  0003865      kidney neoplasm  C63.2        Malignant neoplasm of scrotum   \n",
       "5  0003865      kidney neoplasm    C58       Malignant neoplasm of placenta   \n",
       "6  0003865      kidney neoplasm    C56                                  NaN   \n",
       "7  0003865      kidney neoplasm    C73  Malignant neoplasm of thyroid gland   \n",
       "8  0003865      kidney neoplasm    C64                                  NaN   \n",
       "9  0003865      kidney neoplasm    D23                                  NaN   \n",
       "\n",
       "  mapping_source_prefix mapping_target_prefix  distance  \n",
       "0                   EFO               ICD10CM         1  \n",
       "1                   EFO               ICD10CM         1  \n",
       "2                   EFO               ICD10CM         1  \n",
       "3                   EFO               ICD10CM         1  \n",
       "4             EFO,MONDO               ICD10CM         3  \n",
       "5             EFO,MONDO               ICD10CM         3  \n",
       "6             EFO,MONDO               ICD10CM         3  \n",
       "7    EFO,MONDO,Orphanet               ICD10CM         3  \n",
       "8                   EFO               ICD10CM         1  \n",
       "9          EFO,Orphanet               ICD10CM         3  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract the EFO id for the match in the future.\n",
    "def find_efo(string):\n",
    "   # return str(re.findall(r'(\\d+)', string)[0])\n",
    "     return re.sub(r'EFO:', '', string)\n",
    "\n",
    "EFO2ICD10['EFO'] = EFO2ICD10['EFO'].apply(find_efo)\n",
    "\n",
    "def find_icd10(string):\n",
    "    return re.sub(r'ICD10CM:', '', string)\n",
    "EFO2ICD10['ICD10'] = EFO2ICD10['ICD10'].apply(find_icd10)\n",
    "\n",
    "#EFO2ICD10.reindex(columns = ['EFO_ID' , 'EFO'] + EFO2ICD10.columns.tolist()[1:-1])\n",
    "EFO2ICD10.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "EFO2CCS = EFO2ICD10\n",
    "ce = CCSEngine(mode=\"dx\")\n",
    "\n",
    "def find_ccs(string):\n",
    "    out = ce.get_ccs(string)\n",
    "    result = json.loads(json.dumps(out, indent = 2))\n",
    "    return result['ccs']\n",
    "\n",
    "EFO2CCS['CCS'] = EFO2CCS['ICD10'].apply(find_ccs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In this EFO-CCS dictionary, we have 7000 ICD10 items. Among them, 2969 are not matched, 4031 are matched to a CCS code. To dive deep, the ICD10 items without a match all have missing (NaN) mapped label.\n",
      "\n",
      "Now one EFO item may be assigned multiple CCS code. So we may need to address this as well.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EFO</th>\n",
       "      <th>label</th>\n",
       "      <th>ICD10</th>\n",
       "      <th>mapped_label</th>\n",
       "      <th>mapping_source_prefix</th>\n",
       "      <th>mapping_target_prefix</th>\n",
       "      <th>distance</th>\n",
       "      <th>CCS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M73</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0009470</td>\n",
       "      <td>soft tissue disease</td>\n",
       "      <td>M79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C63.2</td>\n",
       "      <td>Malignant neoplasm of scrotum</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C58</td>\n",
       "      <td>Malignant neoplasm of placenta</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C73</td>\n",
       "      <td>Malignant neoplasm of thyroid gland</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>C64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>0003865</td>\n",
       "      <td>kidney neoplasm</td>\n",
       "      <td>D23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       EFO                label  ICD10                         mapped_label  \\\n",
       "0  0009470  soft tissue disease    M73                                  NaN   \n",
       "1  0009470  soft tissue disease    M75                                  NaN   \n",
       "2  0009470  soft tissue disease    M70                                  NaN   \n",
       "3  0009470  soft tissue disease    M79                                  NaN   \n",
       "4  0003865      kidney neoplasm  C63.2        Malignant neoplasm of scrotum   \n",
       "5  0003865      kidney neoplasm    C58       Malignant neoplasm of placenta   \n",
       "6  0003865      kidney neoplasm    C56                                  NaN   \n",
       "7  0003865      kidney neoplasm    C73  Malignant neoplasm of thyroid gland   \n",
       "8  0003865      kidney neoplasm    C64                                  NaN   \n",
       "9  0003865      kidney neoplasm    D23                                  NaN   \n",
       "\n",
       "  mapping_source_prefix mapping_target_prefix  distance CCS  \n",
       "0                   EFO               ICD10CM         1  na  \n",
       "1                   EFO               ICD10CM         1  na  \n",
       "2                   EFO               ICD10CM         1  na  \n",
       "3                   EFO               ICD10CM         1  na  \n",
       "4             EFO,MONDO               ICD10CM         3  31  \n",
       "5             EFO,MONDO               ICD10CM         3  28  \n",
       "6             EFO,MONDO               ICD10CM         3  na  \n",
       "7    EFO,MONDO,Orphanet               ICD10CM         3  36  \n",
       "8                   EFO               ICD10CM         1  na  \n",
       "9          EFO,Orphanet               ICD10CM         3  na  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('In this EFO-CCS dictionary, we have ' + str(EFO2CCS.shape[0]) + ' ICD10 items. Among them, ' + \n",
    "       str(len(EFO2CCS[EFO2CCS['CCS'] == 'na'])) + ' are not matched, ' + str(len(EFO2CCS[EFO2CCS['CCS'] != 'na'])) +  \n",
    "      ' are matched to a CCS code. ' + 'To dive deep, the ICD10 items without a match all have missing (NaN) mapped label.')\n",
    "\n",
    "print('\\nNow one EFO item may be assigned multiple CCS code. So we may need to address this as well.')\n",
    "\n",
    "EFO2CCS.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color = 'blue'> Tranform the EFO2CCS dictionary into wide format, thus, EFO is able to be an unique identifer. </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EFO</th>\n",
       "      <th>label</th>\n",
       "      <th>ICD10</th>\n",
       "      <th>mapped_label</th>\n",
       "      <th>mapping_source_prefix</th>\n",
       "      <th>mapping_target_prefix</th>\n",
       "      <th>distance</th>\n",
       "      <th>CCS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5216</td>\n",
       "      <td>0000094</td>\n",
       "      <td>B-cell acute lymphoblastic leukemia</td>\n",
       "      <td>C83.70</td>\n",
       "      <td>Burkitt lymphoma, unspecified site</td>\n",
       "      <td>EFO,BAO,CLO,DOID</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5213</td>\n",
       "      <td>0000094</td>\n",
       "      <td>B-cell acute lymphoblastic leukemia</td>\n",
       "      <td>C83.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5214</td>\n",
       "      <td>0000094</td>\n",
       "      <td>B-cell acute lymphoblastic leukemia</td>\n",
       "      <td>C83.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,BAO,CLO,DOID</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5215</td>\n",
       "      <td>0000094</td>\n",
       "      <td>B-cell acute lymphoblastic leukemia</td>\n",
       "      <td>C91.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5570</td>\n",
       "      <td>0000095</td>\n",
       "      <td>chronic lymphocytic leukemia</td>\n",
       "      <td>C91.10</td>\n",
       "      <td>Chronic lymphocytic leukemia of B-cell type no...</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5569</td>\n",
       "      <td>0000095</td>\n",
       "      <td>chronic lymphocytic leukemia</td>\n",
       "      <td>C91.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2332</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2328</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2327</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2326</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2333</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2325</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2329</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2331</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2330</td>\n",
       "      <td>0000096</td>\n",
       "      <td>neoplasm of mature B-cells</td>\n",
       "      <td>C82.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO,MONDO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>298</td>\n",
       "      <td>0000174</td>\n",
       "      <td>Ewing sarcoma</td>\n",
       "      <td>C71.9</td>\n",
       "      <td>Malignant neoplasm of brain, unspecified</td>\n",
       "      <td>EFO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>297</td>\n",
       "      <td>0000174</td>\n",
       "      <td>Ewing sarcoma</td>\n",
       "      <td>C49.6</td>\n",
       "      <td>Malignant neoplasm of connective and soft tiss...</td>\n",
       "      <td>EFO,Orphanet</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>296</td>\n",
       "      <td>0000174</td>\n",
       "      <td>Ewing sarcoma</td>\n",
       "      <td>C40.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>295</td>\n",
       "      <td>0000174</td>\n",
       "      <td>Ewing sarcoma</td>\n",
       "      <td>C40.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>294</td>\n",
       "      <td>0000174</td>\n",
       "      <td>Ewing sarcoma</td>\n",
       "      <td>C40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO</td>\n",
       "      <td>ICD10CM</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          EFO                                label   ICD10  \\\n",
       "5216  0000094  B-cell acute lymphoblastic leukemia  C83.70   \n",
       "5213  0000094  B-cell acute lymphoblastic leukemia   C83.5   \n",
       "5214  0000094  B-cell acute lymphoblastic leukemia   C83.7   \n",
       "5215  0000094  B-cell acute lymphoblastic leukemia   C91.0   \n",
       "5570  0000095         chronic lymphocytic leukemia  C91.10   \n",
       "5569  0000095         chronic lymphocytic leukemia   C91.1   \n",
       "2332  0000096           neoplasm of mature B-cells   C82.0   \n",
       "2328  0000096           neoplasm of mature B-cells   C82.1   \n",
       "2327  0000096           neoplasm of mature B-cells   C82.7   \n",
       "2326  0000096           neoplasm of mature B-cells   C82.3   \n",
       "2333  0000096           neoplasm of mature B-cells   C82.2   \n",
       "2325  0000096           neoplasm of mature B-cells   C82.5   \n",
       "2329  0000096           neoplasm of mature B-cells   C82.4   \n",
       "2331  0000096           neoplasm of mature B-cells   C82.6   \n",
       "2330  0000096           neoplasm of mature B-cells   C82.9   \n",
       "298   0000174                        Ewing sarcoma   C71.9   \n",
       "297   0000174                        Ewing sarcoma   C49.6   \n",
       "296   0000174                        Ewing sarcoma   C40.3   \n",
       "295   0000174                        Ewing sarcoma   C40.2   \n",
       "294   0000174                        Ewing sarcoma   C40.0   \n",
       "\n",
       "                                           mapped_label mapping_source_prefix  \\\n",
       "5216                 Burkitt lymphoma, unspecified site      EFO,BAO,CLO,DOID   \n",
       "5213                                                NaN             EFO,MONDO   \n",
       "5214                                                NaN      EFO,BAO,CLO,DOID   \n",
       "5215                                                NaN             EFO,MONDO   \n",
       "5570  Chronic lymphocytic leukemia of B-cell type no...                   EFO   \n",
       "5569                                                NaN                   EFO   \n",
       "2332                                                NaN    EFO,MONDO,Orphanet   \n",
       "2328                                                NaN    EFO,MONDO,Orphanet   \n",
       "2327                                                NaN    EFO,MONDO,Orphanet   \n",
       "2326                                                NaN    EFO,MONDO,Orphanet   \n",
       "2333                                                NaN    EFO,MONDO,Orphanet   \n",
       "2325                                                NaN    EFO,MONDO,Orphanet   \n",
       "2329                                                NaN    EFO,MONDO,Orphanet   \n",
       "2331                                                NaN    EFO,MONDO,Orphanet   \n",
       "2330                                                NaN    EFO,MONDO,Orphanet   \n",
       "298            Malignant neoplasm of brain, unspecified          EFO,Orphanet   \n",
       "297   Malignant neoplasm of connective and soft tiss...          EFO,Orphanet   \n",
       "296                                                 NaN                   EFO   \n",
       "295                                                 NaN                   EFO   \n",
       "294                                                 NaN                   EFO   \n",
       "\n",
       "     mapping_target_prefix  distance CCS  \n",
       "5216               ICD10CM         3  38  \n",
       "5213               ICD10CM         3  na  \n",
       "5214               ICD10CM         3  na  \n",
       "5215               ICD10CM         3  na  \n",
       "5570               ICD10CM         1  39  \n",
       "5569               ICD10CM         1  na  \n",
       "2332               ICD10CM         3  na  \n",
       "2328               ICD10CM         3  na  \n",
       "2327               ICD10CM         3  na  \n",
       "2326               ICD10CM         3  na  \n",
       "2333               ICD10CM         3  na  \n",
       "2325               ICD10CM         3  na  \n",
       "2329               ICD10CM         3  na  \n",
       "2331               ICD10CM         3  na  \n",
       "2330               ICD10CM         3  na  \n",
       "298                ICD10CM         3  35  \n",
       "297                ICD10CM         3  21  \n",
       "296                ICD10CM         1  na  \n",
       "295                ICD10CM         1  na  \n",
       "294                ICD10CM         1  na  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFO2CCS.sort_values('EFO').head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CCS_LINE</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EFO</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0000094</td>\n",
       "      <td>na</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0000095</td>\n",
       "      <td>na</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0000096</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0000174</td>\n",
       "      <td>na</td>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0000178</td>\n",
       "      <td>13</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CCS_LINE  1    2    3    4    5    6    7    8    9    10   11   12   13   14  \\\n",
       "EFO                                                                             \n",
       "0000094   na   38  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "0000095   na   39  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "0000096   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "0000174   na   21   35  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "0000178   13   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "\n",
       "CCS_LINE   15   16  \n",
       "EFO                 \n",
       "0000094   NaN  NaN  \n",
       "0000095   NaN  NaN  \n",
       "0000096   NaN  NaN  \n",
       "0000174   NaN  NaN  \n",
       "0000178   NaN  NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFO2CCS_long = EFO2CCS[['EFO', 'CCS']].drop_duplicates()\n",
    "EFO2CCS_long['No'] = 1 \n",
    "EFO2CCS_long['CCS_LINE'] = EFO2CCS_long.groupby('EFO')['No'].cumsum()\n",
    "EFO2CCS_long = EFO2CCS_long.pivot(index='EFO', columns='CCS_LINE', values='CCS')\n",
    "#EFO2CCS_long.columns = EFO2CCS_long.columns.droplevel(0)\n",
    "#EFO2CCS_long.columns = ['CCS_'.join(col).strip() for col in EFO2CCS_long.columns.values]\n",
    "EFO2CCS_long.sort_values('EFO').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CCS_LINE</th>\n",
       "      <th>EFO</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0000094</td>\n",
       "      <td>na</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0000095</td>\n",
       "      <td>na</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0000096</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0000174</td>\n",
       "      <td>na</td>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0000178</td>\n",
       "      <td>13</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CCS_LINE      EFO   1    2    3    4    5    6    7    8    9   10   11   12  \\\n",
       "0         0000094  na   38  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "1         0000095  na   39  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "2         0000096  na  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "3         0000174  na   21   35  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "4         0000178  13   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   \n",
       "\n",
       "CCS_LINE   13   14   15   16  \n",
       "0         NaN  NaN  NaN  NaN  \n",
       "1         NaN  NaN  NaN  NaN  \n",
       "2         NaN  NaN  NaN  NaN  \n",
       "3         NaN  NaN  NaN  NaN  \n",
       "4         NaN  NaN  NaN  NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFO2CCS_long.reset_index(inplace=True)\n",
    "EFO2CCS_long.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CCS_LINE</th>\n",
       "      <th>EFO</th>\n",
       "      <th>CCS1</th>\n",
       "      <th>CCS2</th>\n",
       "      <th>CCS3</th>\n",
       "      <th>CCS4</th>\n",
       "      <th>CCS5</th>\n",
       "      <th>CCS6</th>\n",
       "      <th>CCS7</th>\n",
       "      <th>CCS8</th>\n",
       "      <th>CCS9</th>\n",
       "      <th>CCS10</th>\n",
       "      <th>CCS11</th>\n",
       "      <th>CCS12</th>\n",
       "      <th>CCS13</th>\n",
       "      <th>CCS14</th>\n",
       "      <th>CCS15</th>\n",
       "      <th>CCS16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0000094</td>\n",
       "      <td>na</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0000095</td>\n",
       "      <td>na</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0000096</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0000174</td>\n",
       "      <td>na</td>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0000178</td>\n",
       "      <td>13</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CCS_LINE      EFO CCS1 CCS2 CCS3 CCS4 CCS5 CCS6 CCS7 CCS8 CCS9 CCS10 CCS11  \\\n",
       "0         0000094   na   38  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "1         0000095   na   39  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "2         0000096   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "3         0000174   na   21   35  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "4         0000178   13   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "\n",
       "CCS_LINE CCS12 CCS13 CCS14 CCS15 CCS16  \n",
       "0          NaN   NaN   NaN   NaN   NaN  \n",
       "1          NaN   NaN   NaN   NaN   NaN  \n",
       "2          NaN   NaN   NaN   NaN   NaN  \n",
       "3          NaN   NaN   NaN   NaN   NaN  \n",
       "4          NaN   NaN   NaN   NaN   NaN  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EFO2CCS_long.reset_index(drop =True)\n",
    "EFO2CCS_long = EFO2CCS_long.add_prefix('CCS')\n",
    "EFO2CCS_long= EFO2CCS_long.rename(columns = {'CCSEFO':'EFO'})\n",
    "EFO2CCS_long.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge the EFO2CCS dictionary to the GWAS catalog?\n",
    "\n",
    "<font color = \"red\">Basic Idea: </font> \n",
    "* Match the \"Cat_Ancestry\" and \"Cat_Studies\" by the \"STUDY ACCESSION\", then the Cat_Ancestry will have the *EFO (Diseae/Trait)* label info.\n",
    "* Match the \"EFO2CCS\" and \"Cat_Ancestry\" by the \"EFO term?\"\n",
    "\n",
    "Check the unique identifier below before merging the \"Cat_Ancestry\" and \"Cat_Studies\" on \"STUDY ACCESSION\"\n",
    "\n",
    "<code>\n",
    "Cat_Ancestry.head(1)\n",
    "Cat_Studies.head(1)\n",
    "</code>    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now let's check if the study-accession is an unique identifier in the cat_studies data:\n",
      "  Wow, that is to say, in the Cat_Study data, the study-accession is an unique identifer in the data.\n",
      "  So we can use the Cat_Study as an dictionary to map the study-accession in the cat-ancestry data to the EFO terms\n"
     ]
    }
   ],
   "source": [
    "# check the unique identifier in the CAT_STUDIES data\n",
    "\n",
    "print('Now let\\'s check if the study-accession is an unique identifier in the cat_studies data:')\n",
    "if Cat_Studies.duplicated(['STUDY ACCESSION']).any() == False:\n",
    "    print('  Wow, that is to say, in the Cat_Study data, the study-accession is an unique identifer in the data.' + '\\n'\n",
    "          '  So we can use the Cat_Study as an dictionary to map the study-accession in the cat-ancestry data to the EFO terms')\n",
    "else:\n",
    "    print(' In the Cat_Study data, the study-accession is NOT an unique identifer in the data.' + '\\n'\n",
    "          ' This may cause a problem when merge the cat_studies data to the cat_ancestry data.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now let's check if the study-accession is an unique identifier in the Broad_Ancestryal_Study data:\n",
      "  Wow, that is to say, in the Broad_Ancestral_Ancestry data, the study-accession is an unique identifer in the data.\n",
      "  So we can use 1:1 merge with the Broad_Ancestral_Studies data on study-accession.\n"
     ]
    }
   ],
   "source": [
    "# check the unique identifier in the CAT_ANCESTRY data\n",
    "\n",
    "print('Now let\\'s check if the study-accession is an unique identifier in the Broad_Ancestryal_Study data:')\n",
    "if Ancestry.duplicated(['STUDY ACCESSION']).any() == False:\n",
    "    print('  Wow, that is to say, in the Broad_Ancestral_Ancestry data, the study-accession is an unique identifer in the data.' + '\\n'\n",
    "          '  So we can use 1:1 merge with the Broad_Ancestral_Studies data on study-accession.')\n",
    "else:\n",
    "    print(' In the Cat_Ancestry data, the study-accession is NOT an unique identifer in the data.' + '\\n'\n",
    "          ' This means, when merging with the Broad_Ancestral_Studies data on study-accession, we need to use the m:1 merge.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'>Now Let's merge the \"Cat_Ancestry\" with the \"Cat_Studies\" on the \"STUDY-ACCESSION\", the merge validation is 1:1 merge </font>. After this action, we should expect that each item in the \"Cat_Ancestry\" should have a EFO Term which can be used to merge with the EFO2CCS dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ancestry = pd.merge(Ancestry, Cat_Studies, on = 'STUDY ACCESSION', \n",
    "                                      how = 'left', validate = '1:1', indicator = 'str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In total there are 7788 Study-Trait-wise observations in the Broad_Ancestral_Study data.\n",
      "Check the merge outcomes: \n",
      "\n",
      "Great! All observations in the Broad_Ancestral_Study data have been assigned to and EFO term.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID_x</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_African</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>...</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>ASSOCIATION COUNT</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "      <th>str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Complement factor H polymorphism in age-relate...</td>\n",
       "      <td>Age-related macular degeneration</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [103611]</td>\n",
       "      <td>1</td>\n",
       "      <td>age-related macular degeneration</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001365</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID_x      DATE_x  initial_African  \\\n",
       "0      GCST000001    15761122  2005-03-10                0   \n",
       "\n",
       "   initial_African Am./Caribbean  initial_Asian  initial_European  \\\n",
       "0                              0              0               146   \n",
       "\n",
       "   initial_Hispanic/Latin American  initial_In Part Not Recorded  \\\n",
       "0                                0                             0   \n",
       "\n",
       "   initial_Other/Mixed  ...  \\\n",
       "0                    0  ...   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Complement factor H polymorphism in age-relate...   \n",
       "\n",
       "                      DISEASE/TRAIT  \\\n",
       "0  Age-related macular degeneration   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  REPLICATION SAMPLE SIZE  \\\n",
       "0  96 European ancestry cases, 50 European ancest...                      N/A   \n",
       "\n",
       "   PLATFORM [SNPS PASSING QC]  ASSOCIATION COUNT  \\\n",
       "0         Affymetrix [103611]                  1   \n",
       "\n",
       "                       MAPPED_TRAIT                      MAPPED_TRAIT_URI  \\\n",
       "0  age-related macular degeneration  http://www.ebi.ac.uk/efo/EFO_0001365   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY   str  \n",
       "0  Genome-wide genotyping array  both  \n",
       "\n",
       "[1 rows x 42 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('In total there are ' + str(Ancestry.shape[0]) + ' Study-Trait-wise observations in the Broad_Ancestral_Study data.'  + '\\n'\n",
    "      + 'Check the merge outcomes: \\n')\n",
    "if Counter(Ancestry['str'].tolist())['both'] == Ancestry.shape[0]:\n",
    "    print('Great! All observations in the Broad_Ancestral_Study data have been assigned to and EFO term.')\n",
    "else:\n",
    "    print('Opps!' + str(Connter(Ancestry['str'].tolist())['left-only']) + ' of the' + str(Ancestry.shape[0]) + '\\n' \n",
    "          + 'observations have been assigned to an EFO term.')\n",
    "\n",
    "Ancestry.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_efo_type2(string):\n",
    "    find = re.findall(r'EFO_\\d+', string)\n",
    "    if find != []:\n",
    "        find = re.findall(r'EFO_\\d+', string)\n",
    "        find= ','.join(x for x in find)\n",
    "    if find == []:\n",
    "        find = None\n",
    "    return find    \n",
    "\n",
    "Ancestry['Multiple_EFO'] = Ancestry['MAPPED_TRAIT_URI'].apply(find_efo_type2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Be careful! One study may have multiple mapped-trait-uri, thus, in the EFO column, there may be a list of more than one \"EFO-xxxxx\"       So, we may need to split the EFO column into multiple columns each one of which should contain only one mapped EFO term.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID_x</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_African</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>...</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>ASSOCIATION COUNT</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "      <th>str</th>\n",
       "      <th>Multiple_EFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Age-related macular degeneration</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [103611]</td>\n",
       "      <td>1</td>\n",
       "      <td>age-related macular degeneration</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001365</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0001365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>744</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>Perlegen [198345]</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0002508</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0002508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GCST000003</td>\n",
       "      <td>16614226</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Obesity</td>\n",
       "      <td>694 individuals from 288 families</td>\n",
       "      <td>3,213 European ancestry obese cases, 6,210 Eur...</td>\n",
       "      <td>Affymetrix [86604]</td>\n",
       "      <td>0</td>\n",
       "      <td>obesity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001073</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0001073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GCST000004</td>\n",
       "      <td>16648850</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>QT interval</td>\n",
       "      <td>100 European ancestry &gt; 445ms female individua...</td>\n",
       "      <td>200 European ancestry &gt; 85th pct female indivi...</td>\n",
       "      <td>Affymetrix [88500]</td>\n",
       "      <td>1</td>\n",
       "      <td>QT interval</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004682</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0004682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GCST000005</td>\n",
       "      <td>17052657</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>267 European ancestry cases, 270 European ance...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Illumina [408803]</td>\n",
       "      <td>3</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0002508</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0002508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>95</td>\n",
       "      <td>GCST000096</td>\n",
       "      <td>17903295</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1345</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Aging traits</td>\n",
       "      <td>1,345 individuals from 330 families</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [70897]</td>\n",
       "      <td>10</td>\n",
       "      <td>age at menopause, aging, exercise test, age at...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004704, http://p...</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0004704,EFO_0004328,EFO_0005056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>96</td>\n",
       "      <td>GCST000097</td>\n",
       "      <td>17903303</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>678</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Coronary artery calcification</td>\n",
       "      <td>Up to 678 individuals</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [70897]</td>\n",
       "      <td>3</td>\n",
       "      <td>coronary artery calcification</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004723</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0004723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97</td>\n",
       "      <td>GCST000098</td>\n",
       "      <td>17903297</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Cognitive test performance</td>\n",
       "      <td>694 European ancestry individuals</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [70897]</td>\n",
       "      <td>14</td>\n",
       "      <td>neuropsychological test</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003926</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0003926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98</td>\n",
       "      <td>GCST000099</td>\n",
       "      <td>17903307</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1222</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Mean forced vital capacity from 2 exams</td>\n",
       "      <td>Up to 1,222 individuals</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [70897]</td>\n",
       "      <td>2</td>\n",
       "      <td>vital capacity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004312</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0004312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>99</td>\n",
       "      <td>GCST000100</td>\n",
       "      <td>17903292</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>883</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Thyroid stimulating hormone levels</td>\n",
       "      <td>883 individuals</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [70897]</td>\n",
       "      <td>3</td>\n",
       "      <td>thyroid stimulating hormone measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004748</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>both</td>\n",
       "      <td>EFO_0004748</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   STUDY ACCESSION  PUBMEDID_x      DATE_x  initial_African  \\\n",
       "0       GCST000001    15761122  2005-03-10                0   \n",
       "1       GCST000002    16252231  2005-09-09                0   \n",
       "2       GCST000003    16614226  2006-04-14                0   \n",
       "3       GCST000004    16648850  2006-04-30                0   \n",
       "4       GCST000005    17052657  2006-09-28                0   \n",
       "..             ...         ...         ...              ...   \n",
       "95      GCST000096    17903295  2007-09-19                0   \n",
       "96      GCST000097    17903303  2007-09-19                0   \n",
       "97      GCST000098    17903297  2007-09-19                0   \n",
       "98      GCST000099    17903307  2007-09-19                0   \n",
       "99      GCST000100    17903292  2007-09-19                0   \n",
       "\n",
       "    initial_African Am./Caribbean  initial_Asian  initial_European  \\\n",
       "0                               0              0               146   \n",
       "1                               0              1               744   \n",
       "2                               0              0                 0   \n",
       "3                               0              0               200   \n",
       "4                               0              0               537   \n",
       "..                            ...            ...               ...   \n",
       "95                              0              0                 0   \n",
       "96                              0              0                 0   \n",
       "97                              0              0               694   \n",
       "98                              0              0                 0   \n",
       "99                              0              0                 0   \n",
       "\n",
       "    initial_Hispanic/Latin American  initial_In Part Not Recorded  \\\n",
       "0                                 0                             0   \n",
       "1                                 0                             0   \n",
       "2                                 0                           694   \n",
       "3                                 0                             0   \n",
       "4                                 0                             0   \n",
       "..                              ...                           ...   \n",
       "95                                0                          1345   \n",
       "96                                0                           678   \n",
       "97                                0                             0   \n",
       "98                                0                          1222   \n",
       "99                                0                           883   \n",
       "\n",
       "    initial_Other/Mixed  ...                            DISEASE/TRAIT  \\\n",
       "0                     0  ...         Age-related macular degeneration   \n",
       "1                   141  ...                      Parkinson's disease   \n",
       "2                     0  ...                                  Obesity   \n",
       "3                     0  ...                              QT interval   \n",
       "4                     0  ...                      Parkinson's disease   \n",
       "..                  ...  ...                                      ...   \n",
       "95                    0  ...                             Aging traits   \n",
       "96                    0  ...            Coronary artery calcification   \n",
       "97                    0  ...               Cognitive test performance   \n",
       "98                    0  ...  Mean forced vital capacity from 2 exams   \n",
       "99                    0  ...       Thyroid stimulating hormone levels   \n",
       "\n",
       "                                  INITIAL SAMPLE SIZE  \\\n",
       "0   96 European ancestry cases, 50 European ancest...   \n",
       "1   381 European ancestry cases, 363 European ance...   \n",
       "2                   694 individuals from 288 families   \n",
       "3   100 European ancestry > 445ms female individua...   \n",
       "4   267 European ancestry cases, 270 European ance...   \n",
       "..                                                ...   \n",
       "95                1,345 individuals from 330 families   \n",
       "96                              Up to 678 individuals   \n",
       "97                  694 European ancestry individuals   \n",
       "98                            Up to 1,222 individuals   \n",
       "99                                    883 individuals   \n",
       "\n",
       "                              REPLICATION SAMPLE SIZE  \\\n",
       "0                                                 N/A   \n",
       "1   269 European ancestry cases, 272 European ance...   \n",
       "2   3,213 European ancestry obese cases, 6,210 Eur...   \n",
       "3   200 European ancestry > 85th pct female indivi...   \n",
       "4                                                 N/A   \n",
       "..                                                ...   \n",
       "95                                                N/A   \n",
       "96                                                N/A   \n",
       "97                                                N/A   \n",
       "98                                                N/A   \n",
       "99                                                N/A   \n",
       "\n",
       "    PLATFORM [SNPS PASSING QC]  ASSOCIATION COUNT  \\\n",
       "0          Affymetrix [103611]                  1   \n",
       "1            Perlegen [198345]                  1   \n",
       "2           Affymetrix [86604]                  0   \n",
       "3           Affymetrix [88500]                  1   \n",
       "4            Illumina [408803]                  3   \n",
       "..                         ...                ...   \n",
       "95          Affymetrix [70897]                 10   \n",
       "96          Affymetrix [70897]                  3   \n",
       "97          Affymetrix [70897]                 14   \n",
       "98          Affymetrix [70897]                  2   \n",
       "99          Affymetrix [70897]                  3   \n",
       "\n",
       "                                         MAPPED_TRAIT  \\\n",
       "0                    age-related macular degeneration   \n",
       "1                                 Parkinson's disease   \n",
       "2                                             obesity   \n",
       "3                                         QT interval   \n",
       "4                                 Parkinson's disease   \n",
       "..                                                ...   \n",
       "95  age at menopause, aging, exercise test, age at...   \n",
       "96                      coronary artery calcification   \n",
       "97                            neuropsychological test   \n",
       "98                                     vital capacity   \n",
       "99            thyroid stimulating hormone measurement   \n",
       "\n",
       "                                     MAPPED_TRAIT_URI  \\\n",
       "0                http://www.ebi.ac.uk/efo/EFO_0001365   \n",
       "1                http://www.ebi.ac.uk/efo/EFO_0002508   \n",
       "2                http://www.ebi.ac.uk/efo/EFO_0001073   \n",
       "3                http://www.ebi.ac.uk/efo/EFO_0004682   \n",
       "4                http://www.ebi.ac.uk/efo/EFO_0002508   \n",
       "..                                                ...   \n",
       "95  http://www.ebi.ac.uk/efo/EFO_0004704, http://p...   \n",
       "96               http://www.ebi.ac.uk/efo/EFO_0004723   \n",
       "97               http://www.ebi.ac.uk/efo/EFO_0003926   \n",
       "98               http://www.ebi.ac.uk/efo/EFO_0004312   \n",
       "99               http://www.ebi.ac.uk/efo/EFO_0004748   \n",
       "\n",
       "           GENOTYPING TECHNOLOGY   str                         Multiple_EFO  \n",
       "0   Genome-wide genotyping array  both                          EFO_0001365  \n",
       "1   Genome-wide genotyping array  both                          EFO_0002508  \n",
       "2   Genome-wide genotyping array  both                          EFO_0001073  \n",
       "3   Genome-wide genotyping array  both                          EFO_0004682  \n",
       "4   Genome-wide genotyping array  both                          EFO_0002508  \n",
       "..                           ...   ...                                  ...  \n",
       "95  Genome-wide genotyping array  both  EFO_0004704,EFO_0004328,EFO_0005056  \n",
       "96  Genome-wide genotyping array  both                          EFO_0004723  \n",
       "97  Genome-wide genotyping array  both                          EFO_0003926  \n",
       "98  Genome-wide genotyping array  both                          EFO_0004312  \n",
       "99  Genome-wide genotyping array  both                          EFO_0004748  \n",
       "\n",
       "[100 rows x 43 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Be careful! One study may have multiple mapped-trait-uri, thus, in the EFO column, there may be a list of more than one \"EFO-xxxxx\" \\\n",
    "      So, we may need to split the EFO column into multiple columns each one of which should contain only one mapped EFO term.')\n",
    "Ancestry.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID_x</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_African</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>...</th>\n",
       "      <th>EFO47</th>\n",
       "      <th>EFO48</th>\n",
       "      <th>EFO49</th>\n",
       "      <th>EFO50</th>\n",
       "      <th>EFO51</th>\n",
       "      <th>EFO52</th>\n",
       "      <th>EFO53</th>\n",
       "      <th>EFO54</th>\n",
       "      <th>EFO55</th>\n",
       "      <th>EFO56</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>744</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GCST000003</td>\n",
       "      <td>16614226</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>694</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GCST000004</td>\n",
       "      <td>16648850</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GCST000005</td>\n",
       "      <td>17052657</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID_x      DATE_x  initial_African  \\\n",
       "0      GCST000001    15761122  2005-03-10                0   \n",
       "1      GCST000002    16252231  2005-09-09                0   \n",
       "2      GCST000003    16614226  2006-04-14                0   \n",
       "3      GCST000004    16648850  2006-04-30                0   \n",
       "4      GCST000005    17052657  2006-09-28                0   \n",
       "\n",
       "   initial_African Am./Caribbean  initial_Asian  initial_European  \\\n",
       "0                              0              0               146   \n",
       "1                              0              1               744   \n",
       "2                              0              0                 0   \n",
       "3                              0              0               200   \n",
       "4                              0              0               537   \n",
       "\n",
       "   initial_Hispanic/Latin American  initial_In Part Not Recorded  \\\n",
       "0                                0                             0   \n",
       "1                                0                             0   \n",
       "2                                0                           694   \n",
       "3                                0                             0   \n",
       "4                                0                             0   \n",
       "\n",
       "   initial_Other/Mixed  ...  EFO47  EFO48  EFO49  EFO50  EFO51  EFO52  EFO53  \\\n",
       "0                    0  ...   None   None   None   None   None   None   None   \n",
       "1                  141  ...   None   None   None   None   None   None   None   \n",
       "2                    0  ...   None   None   None   None   None   None   None   \n",
       "3                    0  ...   None   None   None   None   None   None   None   \n",
       "4                    0  ...   None   None   None   None   None   None   None   \n",
       "\n",
       "   EFO54  EFO55  EFO56  \n",
       "0   None   None   None  \n",
       "1   None   None   None  \n",
       "2   None   None   None  \n",
       "3   None   None   None  \n",
       "4   None   None   None  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry = Ancestry.join(Ancestry['Multiple_EFO'].str.split(',', expand = True).add_prefix('EFO'))\n",
    "Ancestry.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color = 'blue' size='4pt'> Reshape the Ancestral_Study data from wide format into long format. Each observation is a Study-EFO pair-wise. </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID_x</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_African</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>...</th>\n",
       "      <th>EFO48</th>\n",
       "      <th>EFO49</th>\n",
       "      <th>EFO50</th>\n",
       "      <th>EFO51</th>\n",
       "      <th>EFO52</th>\n",
       "      <th>EFO53</th>\n",
       "      <th>EFO54</th>\n",
       "      <th>EFO55</th>\n",
       "      <th>EFO56</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID_x      DATE_x  initial_African  \\\n",
       "0      GCST000001    15761122  2005-03-10                0   \n",
       "\n",
       "   initial_African Am./Caribbean  initial_Asian  initial_European  \\\n",
       "0                              0              0               146   \n",
       "\n",
       "   initial_Hispanic/Latin American  initial_In Part Not Recorded  \\\n",
       "0                                0                             0   \n",
       "\n",
       "   initial_Other/Mixed  ...  EFO48  EFO49  EFO50  EFO51  EFO52  EFO53  EFO54  \\\n",
       "0                    0  ...   None   None   None   None   None   None   None   \n",
       "\n",
       "   EFO55  EFO56  ID  \n",
       "0   None   None   0  \n",
       "\n",
       "[1 rows x 101 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FIRST, create unique id for each observation for reshape.\n",
    "Ancestry['ID'] = Ancestry.index\n",
    "Ancestry.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The long-format Ancestral_Study_long data has 9927 Observations with non-missing EFO terms.\n",
      "However, look at the EFO columns, the format is not suitable for merging to the EFO2CCS dictionary. So we address this now.\n",
      "We need to transform: EFO_0001365 --> 0001365\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Funded_AHRQ</th>\n",
       "      <th>Funded</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>Multiple_EFO</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>replication_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>...</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>replication_African</th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>replication_In Part Not Recorded</th>\n",
       "      <th>replication_African Am./Caribbean</th>\n",
       "      <th>Funding_merge</th>\n",
       "      <th>PUBMEDID_x</th>\n",
       "      <th>DATE_y</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>EFO</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th>EFO_LINE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0001365</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-06-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>15761122</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>Science</td>\n",
       "      <td>0001365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0002508</td>\n",
       "      <td>744</td>\n",
       "      <td>1</td>\n",
       "      <td>541</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-06-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>16252231</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>0002508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>694</td>\n",
       "      <td>EFO_0001073</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10527</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>694 individuals from 288 families</td>\n",
       "      <td>0</td>\n",
       "      <td>2009-02-14</td>\n",
       "      <td>229</td>\n",
       "      <td>866</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>16614226</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>Science</td>\n",
       "      <td>0001073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0004682</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>4851</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>100 European ancestry &gt; 445ms female individua...</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-06-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>16648850</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>0004682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0002508</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>267 European ancestry cases, 270 European ance...</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-06-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>17052657</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>Lancet Neurol</td>\n",
       "      <td>0002508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0003926</td>\n",
       "      <td>694</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>694 European ancestry individuals</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-09-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>17903297</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>0003926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>1222</td>\n",
       "      <td>EFO_0004312</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Up to 1,222 individuals</td>\n",
       "      <td>0</td>\n",
       "      <td>2009-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>17903307</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>0004312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>99</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>883</td>\n",
       "      <td>EFO_0004748</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>883 individuals</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-09-02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>17903292</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>0004748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>1094</td>\n",
       "      <td>EFO_0004685</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Up to 1,094 individuals from 241 families</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-09-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>17903296</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>0004685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0004298</td>\n",
       "      <td>1238</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Up to 1,238 European ancestry individuals</td>\n",
       "      <td>0</td>\n",
       "      <td>2008-09-17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>funding_info_y</td>\n",
       "      <td>17903301</td>\n",
       "      <td>2007-09-19</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>0004298</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Funded_AHRQ  Funded  initial_African Am./Caribbean      DATE_x  \\\n",
       "ID  EFO_LINE                                                                   \n",
       "0   0                 0.0     6.0                              0  2005-03-10   \n",
       "1   0                 0.0     8.0                              0  2005-09-09   \n",
       "2   0                 0.0    11.0                              0  2006-04-14   \n",
       "3   0                 0.0     3.0                              0  2006-04-30   \n",
       "4   0                 0.0     2.0                              0  2006-09-28   \n",
       "...                   ...     ...                            ...         ...   \n",
       "97  0                 0.0     9.0                              0  2007-09-19   \n",
       "98  0                 0.0     3.0                              0  2007-09-19   \n",
       "99  0                 0.0     3.0                              0  2007-09-19   \n",
       "100 0                 0.0     7.0                              0  2007-09-19   \n",
       "101 0                 0.0    17.0                              0  2007-09-19   \n",
       "\n",
       "              initial_In Part Not Recorded Multiple_EFO  initial_European  \\\n",
       "ID  EFO_LINE                                                                \n",
       "0   0                                    0  EFO_0001365               146   \n",
       "1   0                                    0  EFO_0002508               744   \n",
       "2   0                                  694  EFO_0001073                 0   \n",
       "3   0                                    0  EFO_0004682               200   \n",
       "4   0                                    0  EFO_0002508               537   \n",
       "...                                    ...          ...               ...   \n",
       "97  0                                    0  EFO_0003926               694   \n",
       "98  0                                 1222  EFO_0004312                 0   \n",
       "99  0                                  883  EFO_0004748                 0   \n",
       "100 0                                 1094  EFO_0004685                 0   \n",
       "101 0                                    0  EFO_0004298              1238   \n",
       "\n",
       "              initial_Asian  replication_European  \\\n",
       "ID  EFO_LINE                                        \n",
       "0   0                     0                     0   \n",
       "1   0                     1                   541   \n",
       "2   0                     0                 10527   \n",
       "3   0                     0                  4851   \n",
       "4   0                     0                     0   \n",
       "...                     ...                   ...   \n",
       "97  0                     0                     0   \n",
       "98  0                     0                     0   \n",
       "99  0                     0                     0   \n",
       "100 0                     0                     0   \n",
       "101 0                     0                     0   \n",
       "\n",
       "              initial_Hispanic/Latin American  ...  \\\n",
       "ID  EFO_LINE                                   ...   \n",
       "0   0                                       0  ...   \n",
       "1   0                                       0  ...   \n",
       "2   0                                       0  ...   \n",
       "3   0                                       0  ...   \n",
       "4   0                                       0  ...   \n",
       "...                                       ...  ...   \n",
       "97  0                                       0  ...   \n",
       "98  0                                       0  ...   \n",
       "99  0                                       0  ...   \n",
       "100 0                                       0  ...   \n",
       "101 0                                       0  ...   \n",
       "\n",
       "                                            INITIAL SAMPLE SIZE  \\\n",
       "ID  EFO_LINE                                                      \n",
       "0   0         96 European ancestry cases, 50 European ancest...   \n",
       "1   0         381 European ancestry cases, 363 European ance...   \n",
       "2   0                         694 individuals from 288 families   \n",
       "3   0         100 European ancestry > 445ms female individua...   \n",
       "4   0         267 European ancestry cases, 270 European ance...   \n",
       "...                                                         ...   \n",
       "97  0                         694 European ancestry individuals   \n",
       "98  0                                   Up to 1,222 individuals   \n",
       "99  0                                           883 individuals   \n",
       "100 0                 Up to 1,094 individuals from 241 families   \n",
       "101 0                 Up to 1,238 European ancestry individuals   \n",
       "\n",
       "             replication_African  DATE ADDED TO CATALOG  \\\n",
       "ID  EFO_LINE                                              \n",
       "0   0                          0             2008-06-16   \n",
       "1   0                          0             2008-06-16   \n",
       "2   0                          0             2009-02-14   \n",
       "3   0                          0             2008-06-16   \n",
       "4   0                          0             2008-06-16   \n",
       "...                          ...                    ...   \n",
       "97  0                          0             2008-09-10   \n",
       "98  0                          0             2009-09-28   \n",
       "99  0                          0             2008-09-02   \n",
       "100 0                          0             2008-09-10   \n",
       "101 0                          0             2008-09-17   \n",
       "\n",
       "              replication_In Part Not Recorded  \\\n",
       "ID  EFO_LINE                                     \n",
       "0   0                                        0   \n",
       "1   0                                        0   \n",
       "2   0                                      229   \n",
       "3   0                                        0   \n",
       "4   0                                        0   \n",
       "...                                        ...   \n",
       "97  0                                        0   \n",
       "98  0                                        0   \n",
       "99  0                                        0   \n",
       "100 0                                        0   \n",
       "101 0                                        0   \n",
       "\n",
       "             replication_African Am./Caribbean   Funding_merge PUBMEDID_x  \\\n",
       "ID  EFO_LINE                                                                \n",
       "0   0                                        0  funding_info_y   15761122   \n",
       "1   0                                        0  funding_info_y   16252231   \n",
       "2   0                                      866  funding_info_y   16614226   \n",
       "3   0                                        0  funding_info_y   16648850   \n",
       "4   0                                        0  funding_info_y   17052657   \n",
       "...                                        ...             ...        ...   \n",
       "97  0                                        0  funding_info_y   17903297   \n",
       "98  0                                        0  funding_info_y   17903307   \n",
       "99  0                                        0  funding_info_y   17903292   \n",
       "100 0                                        0  funding_info_y   17903296   \n",
       "101 0                                        0  funding_info_y   17903301   \n",
       "\n",
       "                  DATE_y         JOURNAL      EFO  \n",
       "ID  EFO_LINE                                       \n",
       "0   0         2005-03-10         Science  0001365  \n",
       "1   0         2005-09-09  Am J Hum Genet  0002508  \n",
       "2   0         2006-04-14         Science  0001073  \n",
       "3   0         2006-04-30       Nat Genet  0004682  \n",
       "4   0         2006-09-28   Lancet Neurol  0002508  \n",
       "...                  ...             ...      ...  \n",
       "97  0         2007-09-19   BMC Med Genet  0003926  \n",
       "98  0         2007-09-19   BMC Med Genet  0004312  \n",
       "99  0         2007-09-19   BMC Med Genet  0004748  \n",
       "100 0         2007-09-19   BMC Med Genet  0004685  \n",
       "101 0         2007-09-19   BMC Med Genet  0004298  \n",
       "\n",
       "[100 rows x 44 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Second, reshape from wide to long, id = 'ID', EFO_LINE = 1,2,3.4...56\n",
    "Ancestry = pd.wide_to_long(Ancestry, stubnames= 'EFO', i = 'ID',  j ='EFO_LINE')\n",
    "# drop the observations when EFO is none\n",
    "Ancestry = Ancestry[Ancestry['EFO'].notnull()]\n",
    "print('The long-format Ancestral_Study_long data has ' + str(Ancestry.shape[0]) + ' Observations with non-missing EFO terms.')\n",
    "print('However, look at the EFO columns, the format is not suitable for merging to the EFO2CCS dictionary. So we address this now.')\n",
    "print('We need to transform: EFO_0001365 --> 0001365')\n",
    "Ancestry['EFO'] = Ancestry['EFO'].str.replace('EFO_', '')\n",
    "Ancestry.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The EFO2CCS dictionary has 2136 observations.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CCS_LINE</th>\n",
       "      <th>EFO</th>\n",
       "      <th>CCS1</th>\n",
       "      <th>CCS2</th>\n",
       "      <th>CCS3</th>\n",
       "      <th>CCS4</th>\n",
       "      <th>CCS5</th>\n",
       "      <th>CCS6</th>\n",
       "      <th>CCS7</th>\n",
       "      <th>CCS8</th>\n",
       "      <th>CCS9</th>\n",
       "      <th>CCS10</th>\n",
       "      <th>CCS11</th>\n",
       "      <th>CCS12</th>\n",
       "      <th>CCS13</th>\n",
       "      <th>CCS14</th>\n",
       "      <th>CCS15</th>\n",
       "      <th>CCS16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0000094</td>\n",
       "      <td>na</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0000095</td>\n",
       "      <td>na</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0000096</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0000174</td>\n",
       "      <td>na</td>\n",
       "      <td>21</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0000178</td>\n",
       "      <td>13</td>\n",
       "      <td>na</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CCS_LINE      EFO CCS1 CCS2 CCS3 CCS4 CCS5 CCS6 CCS7 CCS8 CCS9 CCS10 CCS11  \\\n",
       "0         0000094   na   38  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "1         0000095   na   39  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "2         0000096   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "3         0000174   na   21   35  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "4         0000178   13   na  NaN  NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   \n",
       "\n",
       "CCS_LINE CCS12 CCS13 CCS14 CCS15 CCS16  \n",
       "0          NaN   NaN   NaN   NaN   NaN  \n",
       "1          NaN   NaN   NaN   NaN   NaN  \n",
       "2          NaN   NaN   NaN   NaN   NaN  \n",
       "3          NaN   NaN   NaN   NaN   NaN  \n",
       "4          NaN   NaN   NaN   NaN   NaN  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('The EFO2CCS dictionary has ' + str(EFO2CCS_long.shape[0]) + ' observations.')\n",
    "EFO2CCS_long.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Each Study-EFO obserbation may be matched to more than one CCS code. This is currently a wide format, For further analysis,       it should be reshape to long format, each observation is a Study-EFO-CCS pair.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Funded_AHRQ</th>\n",
       "      <th>Funded</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>Multiple_EFO</th>\n",
       "      <th>initial_European</th>\n",
       "      <th>initial_Asian</th>\n",
       "      <th>replication_European</th>\n",
       "      <th>initial_Hispanic/Latin American</th>\n",
       "      <th>...</th>\n",
       "      <th>CCS8</th>\n",
       "      <th>CCS9</th>\n",
       "      <th>CCS10</th>\n",
       "      <th>CCS11</th>\n",
       "      <th>CCS12</th>\n",
       "      <th>CCS13</th>\n",
       "      <th>CCS14</th>\n",
       "      <th>CCS15</th>\n",
       "      <th>CCS16</th>\n",
       "      <th>CCS_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0001365</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0002508</td>\n",
       "      <td>744</td>\n",
       "      <td>1</td>\n",
       "      <td>541</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>694</td>\n",
       "      <td>EFO_0001073</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10527</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-04-30</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0004682</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>4851</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0002508</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Funded_AHRQ  Funded  initial_African Am./Caribbean      DATE_x  \\\n",
       "0          0.0     6.0                              0  2005-03-10   \n",
       "1          0.0     8.0                              0  2005-09-09   \n",
       "2          0.0    11.0                              0  2006-04-14   \n",
       "3          0.0     3.0                              0  2006-04-30   \n",
       "4          0.0     2.0                              0  2006-09-28   \n",
       "\n",
       "   initial_In Part Not Recorded Multiple_EFO  initial_European  initial_Asian  \\\n",
       "0                             0  EFO_0001365               146              0   \n",
       "1                             0  EFO_0002508               744              1   \n",
       "2                           694  EFO_0001073                 0              0   \n",
       "3                             0  EFO_0004682               200              0   \n",
       "4                             0  EFO_0002508               537              0   \n",
       "\n",
       "   replication_European  initial_Hispanic/Latin American  ... CCS8 CCS9  \\\n",
       "0                     0                                0  ...  NaN  NaN   \n",
       "1                   541                                0  ...  NaN  NaN   \n",
       "2                 10527                                0  ...  NaN  NaN   \n",
       "3                  4851                                0  ...  NaN  NaN   \n",
       "4                     0                                0  ...  NaN  NaN   \n",
       "\n",
       "   CCS10  CCS11 CCS12 CCS13 CCS14  CCS15 CCS16  CCS_merge  \n",
       "0    NaN    NaN   NaN   NaN   NaN    NaN   NaN       both  \n",
       "1    NaN    NaN   NaN   NaN   NaN    NaN   NaN       both  \n",
       "2    NaN    NaN   NaN   NaN   NaN    NaN   NaN       both  \n",
       "3    NaN    NaN   NaN   NaN   NaN    NaN   NaN  left_only  \n",
       "4    NaN    NaN   NaN   NaN   NaN    NaN   NaN       both  \n",
       "\n",
       "[5 rows x 61 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry = pd.merge(Ancestry, EFO2CCS_long, on = 'EFO',how = 'left',  indicator = 'CCS_merge', validate = 'm:1')\n",
    "print('Each Study-EFO obserbation may be matched to more than one CCS code. This is currently a wide format, For further analysis, \\\n",
    "      it should be reshape to long format, each observation is a Study-EFO-CCS pair.')\n",
    "Ancestry.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ancestry['ID'] = Ancestry.index\n",
    "Ancestry = pd.wide_to_long(Ancestry, stubnames= 'CCS', i = 'ID',  j ='CCS_LINE')\n",
    "Ancestry = Ancestry.reset_index(level = 1, drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In the Ancestral_Study_CCS data, each observation is a Study-EFO-CCS pair. That is to say, a study may be assigned to       multiple EFO terms, and a typical EFO term may be matched to multiple CCS code. Therefore, in the Ancestral_Study_CCS       data, a study may appear more than once, and meanwhile, a EFO term may appear more than once.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replication_Hispanic/Latin American</th>\n",
       "      <th>Funded_AHRQ</th>\n",
       "      <th>Funded</th>\n",
       "      <th>initial_African Am./Caribbean</th>\n",
       "      <th>DATE_x</th>\n",
       "      <th>initial_N</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>initial_In Part Not Recorded</th>\n",
       "      <th>Multiple_EFO</th>\n",
       "      <th>initial_Other/Mixed</th>\n",
       "      <th>...</th>\n",
       "      <th>ASSOCIATION COUNT</th>\n",
       "      <th>Funded_US</th>\n",
       "      <th>DATE_y</th>\n",
       "      <th>Funded_NIH</th>\n",
       "      <th>EFO</th>\n",
       "      <th>LINK</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>str</th>\n",
       "      <th>Funded_ARRA</th>\n",
       "      <th>CCS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>146</td>\n",
       "      <td>N/A</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0001365</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0001365</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/15761122</td>\n",
       "      <td>Science</td>\n",
       "      <td>both</td>\n",
       "      <td>6.0</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>886</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0002508</td>\n",
       "      <td>141</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0002508</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/16252231</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>both</td>\n",
       "      <td>8.0</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>694</td>\n",
       "      <td>3,213 European ancestry obese cases, 6,210 Eur...</td>\n",
       "      <td>694</td>\n",
       "      <td>EFO_0001073</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2006-04-14</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0001073</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/16614226</td>\n",
       "      <td>Science</td>\n",
       "      <td>both</td>\n",
       "      <td>11.0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>537</td>\n",
       "      <td>N/A</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0002508</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2006-09-28</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0002508</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/17052657</td>\n",
       "      <td>Lancet Neurol</td>\n",
       "      <td>both</td>\n",
       "      <td>1.0</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2006-10-19</td>\n",
       "      <td>226</td>\n",
       "      <td>N/A</td>\n",
       "      <td>0</td>\n",
       "      <td>EFO_0001365</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2006-10-19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0001365</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/17053108</td>\n",
       "      <td>Science</td>\n",
       "      <td>both</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    replication_Hispanic/Latin American  Funded_AHRQ  Funded  \\\n",
       "ID                                                             \n",
       "0                                     0          0.0     6.0   \n",
       "1                                     0          0.0     8.0   \n",
       "2                                     0          0.0    11.0   \n",
       "4                                     0          0.0     2.0   \n",
       "5                                     0          0.0     0.0   \n",
       "\n",
       "    initial_African Am./Caribbean      DATE_x  initial_N  \\\n",
       "ID                                                         \n",
       "0                               0  2005-03-10        146   \n",
       "1                               0  2005-09-09        886   \n",
       "2                               0  2006-04-14        694   \n",
       "4                               0  2006-09-28        537   \n",
       "5                               0  2006-10-19        226   \n",
       "\n",
       "                              REPLICATION SAMPLE SIZE  \\\n",
       "ID                                                      \n",
       "0                                                 N/A   \n",
       "1   269 European ancestry cases, 272 European ance...   \n",
       "2   3,213 European ancestry obese cases, 6,210 Eur...   \n",
       "4                                                 N/A   \n",
       "5                                                 N/A   \n",
       "\n",
       "    initial_In Part Not Recorded Multiple_EFO  initial_Other/Mixed  ...  \\\n",
       "ID                                                                  ...   \n",
       "0                              0  EFO_0001365                    0  ...   \n",
       "1                              0  EFO_0002508                  141  ...   \n",
       "2                            694  EFO_0001073                    0  ...   \n",
       "4                              0  EFO_0002508                    0  ...   \n",
       "5                              0  EFO_0001365                    0  ...   \n",
       "\n",
       "    ASSOCIATION COUNT  Funded_US      DATE_y  Funded_NIH      EFO  \\\n",
       "ID                                                                  \n",
       "0                   1        6.0  2005-03-10         6.0  0001365   \n",
       "1                   1        8.0  2005-09-09         8.0  0002508   \n",
       "2                   0       11.0  2006-04-14        11.0  0001073   \n",
       "4                   3        1.0  2006-09-28         1.0  0002508   \n",
       "5                   1        0.0  2006-10-19         0.0  0001365   \n",
       "\n",
       "                                    LINK         JOURNAL   str Funded_ARRA  \\\n",
       "ID                                                                           \n",
       "0   www.ncbi.nlm.nih.gov/pubmed/15761122         Science  both         6.0   \n",
       "1   www.ncbi.nlm.nih.gov/pubmed/16252231  Am J Hum Genet  both         8.0   \n",
       "2   www.ncbi.nlm.nih.gov/pubmed/16614226         Science  both        11.0   \n",
       "4   www.ncbi.nlm.nih.gov/pubmed/17052657   Lancet Neurol  both         1.0   \n",
       "5   www.ncbi.nlm.nih.gov/pubmed/17053108         Science  both         0.0   \n",
       "\n",
       "    CCS  \n",
       "ID       \n",
       "0    95  \n",
       "1    79  \n",
       "2    58  \n",
       "4    79  \n",
       "5    95  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry = Ancestry[Ancestry['CCS'].notnull()]\n",
    "Ancestry = Ancestry[Ancestry['CCS'] != 'na']\n",
    "print('In the Ancestral_Study_CCS data, each observation is a Study-EFO-CCS pair. That is to say, a study may be assigned to \\\n",
    "      multiple EFO terms, and a typical EFO term may be matched to multiple CCS code. Therefore, in the Ancestral_Study_CCS \\\n",
    "      data, a study may appear more than once, and meanwhile, a EFO term may appear more than once.')\n",
    "Ancestry.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['replication_Hispanic/Latin American', 'Funded_AHRQ', 'Funded',\n",
       "       'initial_African Am./Caribbean', 'DATE_x', 'initial_N',\n",
       "       'REPLICATION SAMPLE SIZE', 'initial_In Part Not Recorded',\n",
       "       'Multiple_EFO', 'initial_Other/Mixed', 'PUBMEDID_y', 'initial_African',\n",
       "       'initial_European', 'initial_Asian', 'replication_European',\n",
       "       'initial_Hispanic/Latin American', 'FIRST AUTHOR', 'STUDY',\n",
       "       'GENOTYPING TECHNOLOGY', 'Funded_HHS', 'replication_N',\n",
       "       'INITIAL SAMPLE SIZE', 'DISEASE/TRAIT', 'replication_African',\n",
       "       'CCS_merge', 'DATE ADDED TO CATALOG',\n",
       "       'replication_In Part Not Recorded', 'PLATFORM [SNPS PASSING QC]',\n",
       "       'MAPPED_TRAIT', 'replication_Other/Mixed',\n",
       "       'replication_African Am./Caribbean', 'STUDY ACCESSION', 'Funding_merge',\n",
       "       'replication_Asian', 'PUBMEDID_x', 'MAPPED_TRAIT_URI',\n",
       "       'ASSOCIATION COUNT', 'Funded_US', 'DATE_y', 'Funded_NIH', 'EFO', 'LINK',\n",
       "       'JOURNAL', 'str', 'Funded_ARRA', 'CCS'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge CCS-Disease Dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCS</th>\n",
       "      <th>Disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Infectious diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>Infectious diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>Infectious diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Infectious diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>Infectious diseases</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CCS              Disease\n",
       "0   1  Infectious diseases\n",
       "1   2  Infectious diseases\n",
       "2   3  Infectious diseases\n",
       "3   4  Infectious diseases\n",
       "4   5  Infectious diseases"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CCS_Disease = pd.read_excel(os.path.abspath(\n",
    "                           os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'MEPS',\n",
    "                                        'Support',\n",
    "                                        'CCS_Disease_Dictionary.xlsx')),\n",
    "                           header=0, encoding='utf-8',\n",
    "                           index_col=False)\n",
    "CCS_Disease['CCS'] = CCS_Disease['CCS'].astype(str)\n",
    "CCS_Disease.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ancestry = pd.merge(Ancestry, CCS_Disease, on = 'CCS', \n",
    "                                      how = 'left', validate = 'm:1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['replication_Hispanic/Latin American', 'Funded_AHRQ', 'Funded',\n",
       "       'initial_African Am./Caribbean', 'DATE_x', 'initial_N',\n",
       "       'REPLICATION SAMPLE SIZE', 'initial_In Part Not Recorded',\n",
       "       'Multiple_EFO', 'initial_Other/Mixed', 'PUBMEDID_y', 'initial_African',\n",
       "       'initial_European', 'initial_Asian', 'replication_European',\n",
       "       'initial_Hispanic/Latin American', 'FIRST AUTHOR', 'STUDY',\n",
       "       'GENOTYPING TECHNOLOGY', 'Funded_HHS', 'replication_N',\n",
       "       'INITIAL SAMPLE SIZE', 'DISEASE/TRAIT', 'replication_African',\n",
       "       'CCS_merge', 'DATE ADDED TO CATALOG',\n",
       "       'replication_In Part Not Recorded', 'PLATFORM [SNPS PASSING QC]',\n",
       "       'MAPPED_TRAIT', 'replication_Other/Mixed',\n",
       "       'replication_African Am./Caribbean', 'STUDY ACCESSION', 'Funding_merge',\n",
       "       'replication_Asian', 'PUBMEDID_x', 'MAPPED_TRAIT_URI',\n",
       "       'ASSOCIATION COUNT', 'Funded_US', 'DATE_y', 'Funded_NIH', 'EFO', 'LINK',\n",
       "       'JOURNAL', 'str', 'Funded_ARRA', 'CCS', 'Disease'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Ancestry.rename(columns = {'initial_African Am./Caribbean':'ini_African_American', \n",
    "                            'initial_Hispanic/Latin American':'ini_Hispanic',\n",
    "                            'initial_Asian': 'ini_Asian',\n",
    "                            'initial_European': 'ini_European',\n",
    "                            'initial_Other/Mixed': 'ini_Mixed',\n",
    "                            'initial_In Part Not Recorded': 'ini_NR',\n",
    "                            'initial_African': 'ini_African',\n",
    "                            'replication_African Am./Caribbean':'rep_African_American', \n",
    "                            'replication_Hispanic/Latin American':'rep_Hispanic',\n",
    "                            'replication_Asian': 'rep_Asian',\n",
    "                            'replication_European': 'rep_European',\n",
    "                            'replication_Other/Mixed': 'rep_Mixed',\n",
    "                            'replication_In Part Not Recorded': 'rep_NR',\n",
    "                            'replication_African': 'rep_African',\n",
    "                            'initial_N':'ini_N',\n",
    "                            'replication_N': 'rep_N'}, inplace = True)\n",
    "\n",
    "\n",
    "\n",
    "Ancestry['N_African'] = Ancestry['ini_African'] + Ancestry['rep_African']\n",
    "Ancestry['N_African_American'] = Ancestry['ini_African_American'] + Ancestry['rep_African_American'] \n",
    "Ancestry['N_Asian'] = Ancestry['ini_Asian'] + Ancestry['rep_Asian'] \n",
    "Ancestry['N_European'] = Ancestry['ini_European'] + Ancestry['rep_European']       \n",
    "Ancestry['N_Hispanic'] = Ancestry['ini_Hispanic'] + Ancestry['rep_Hispanic']  \n",
    "Ancestry['N_Mixed'] = Ancestry['ini_Mixed']  + Ancestry['rep_Mixed']\n",
    "Ancestry['N_NR'] = Ancestry['ini_NR']  + Ancestry['rep_NR']\n",
    "\n",
    "Ancestry['N'] = Ancestry['ini_N'] + Ancestry['rep_N']  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ancestry.to_csv(os.path.abspath(os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'Catalogue',\n",
    "                                        'Synthetic',\n",
    "                                        'Ancestral_Study_CCS.csv')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find the ARRA Supported Studies (2009-2010)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the ARRA grant list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARRA = pd.read_excel(os.path.abspath(os.path.join('__file__',\n",
    "                                                         '../..',\n",
    "                                                         'data',\n",
    "                                                         'Support',\n",
    "                                                         'NIH ARRA Grant List',\n",
    "                                                         'ARRA-projects.xls')), \n",
    "                             sheet_name = \"ARRA\", header=0, encoding='utf-8',\n",
    "                             index_col=False)\n",
    "\n",
    "ARRA.columns = ARRA.columns.str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21712, 25)"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARRA.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARRA['Project Number'] = ARRA['Project Number'].str.strip()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the GWAS study grant info list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>Agency</th>\n",
       "      <th>GrantCountry</th>\n",
       "      <th>GrantID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>24029427</td>\n",
       "      <td>NIDDK NIH HHS</td>\n",
       "      <td>United States</td>\n",
       "      <td>U01DK057303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>24029427</td>\n",
       "      <td>NIDDK NIH HHS</td>\n",
       "      <td>United States</td>\n",
       "      <td>U01DK057304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>24029427</td>\n",
       "      <td>Medical Research Council</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>MR/K003364/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>24029427</td>\n",
       "      <td>Wellcome Trust</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>WT090355/A/09/Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>24029427</td>\n",
       "      <td>NCRR NIH HHS</td>\n",
       "      <td>United States</td>\n",
       "      <td>M01-RR-01346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PUBMEDID                    Agency    GrantCountry          GrantID\n",
       "0  24029427             NIDDK NIH HHS   United States      U01DK057303\n",
       "1  24029427             NIDDK NIH HHS   United States      U01DK057304\n",
       "2  24029427  Medical Research Council  United Kingdom     MR/K003364/1\n",
       "3  24029427            Wellcome Trust  United Kingdom  WT090355/A/09/Z\n",
       "4  24029427              NCRR NIH HHS   United States     M01-RR-01346"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWAS_Grant = pd.read_csv(os.path.abspath(os.path.join('__file__',\n",
    "                                                         '../..',\n",
    "                                                         'data',\n",
    "                                                         'PUBMED',\n",
    "                                                         'Pubmed_FunderInfo.csv')),\n",
    "                             header=0, encoding='utf-8',\n",
    "                             index_col=False, sep =',')\n",
    "\n",
    "GWAS_Grant.columns = GWAS_Grant.columns.str.strip()\n",
    "GWAS_Grant['GrantID'] = GWAS_Grant['GrantID'].str.replace(' ', '')\n",
    "GWAS_Grant['GrantID'] = GWAS_Grant['GrantID'].fillna('NOT REPORTED')\n",
    "GWAS_Grant.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Example of **fuzzywuzzy**:\n",
    "<code>\n",
    "fuzzResult = process.extract(\"R01CA082659\", choice, scorer = fuzz.partial_ratio, limit = 5) \n",
    "fuzzResult\n",
    "[('3R01CA082659-11S1', 100, 11437),\n",
    " ('3R01CA018689-33S1', 82, 779),\n",
    " ('3R01CA018689-33S2', 82, 780),\n",
    " ('2R01CA085419-10', 82, 2332),\n",
    " ('5R01CA085419-11', 82, 2352)]\n",
    "</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10848, 1)"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "choice = ARRA['Project Number']\n",
    "GWAS_Grant = GWAS_Grant[GWAS_Grant['GrantCountry'] == 'United States'][['GrantID']].drop_duplicates()\n",
    "GWAS_Grant.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 8h 28min 25s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for index, row in GWAS_Grant.iterrows():\n",
    "    fuzzResult = process.extract((row['GrantID']), choice , scorer = fuzz.partial_ratio, limit = 3)\n",
    "    GWAS_Grant.loc[index, 'ARRA_MatchedGrant1'] = fuzzResult[0][0]\n",
    "    GWAS_Grant.loc[index, 'ARRA_MatchedScore1'] = fuzzResult[0][1]\n",
    "    GWAS_Grant.loc[index, 'ARRA_MatchedGrant2'] = fuzzResult[1][0]\n",
    "    GWAS_Grant.loc[index, 'ARRA_MatchedScore2'] = fuzzResult[1][1]\n",
    "    GWAS_Grant.loc[index, 'ARRA_MatchedGrant3'] = fuzzResult[2][0]\n",
    "    GWAS_Grant.loc[index, 'ARRA_MatchedScore3'] = fuzzResult[2][1]\n",
    "    \n",
    "GWAS_Grant.to_csv(os.path.abspath(os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'Catalogue',\n",
    "                                        'Synthetic',\n",
    "                                        'fuzz.csv')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Match the GWAS Grant Info data with the ARRA Grant Detailed info via the Matched GrantID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GrantID</th>\n",
       "      <th>ARRA_MatchedGrant1</th>\n",
       "      <th>ARRA_MatchedScore1</th>\n",
       "      <th>ARRA_MatchedGrant2</th>\n",
       "      <th>ARRA_MatchedScore2</th>\n",
       "      <th>ARRA_MatchedGrant3</th>\n",
       "      <th>ARRA_MatchedScore3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>U01DK057303</td>\n",
       "      <td>3U01DK061730-08S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK061730-09S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK062503-08S1</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>U01DK057304</td>\n",
       "      <td>3U01DK061734-09S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK072504-05S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK057149-11S1</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M01-RR-01346</td>\n",
       "      <td>3R01RR014166-16S1</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3R01RR021746-03S1</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3R01RR023168-03S1</td>\n",
       "      <td>58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>U01DK57304</td>\n",
       "      <td>3U01DK057149-11S1</td>\n",
       "      <td>80.0</td>\n",
       "      <td>3R01DK057030-08S1</td>\n",
       "      <td>80.0</td>\n",
       "      <td>3U01DK057131-11S1</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>U01DK057298</td>\n",
       "      <td>3U01DK057219-12S1</td>\n",
       "      <td>91.0</td>\n",
       "      <td>3U01DK061728-09S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK072493-04S1</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GrantID ARRA_MatchedGrant1  ARRA_MatchedScore1 ARRA_MatchedGrant2  \\\n",
       "0   U01DK057303  3U01DK061730-08S1                82.0  3U01DK061730-09S1   \n",
       "1   U01DK057304  3U01DK061734-09S1                82.0  3U01DK072504-05S1   \n",
       "4  M01-RR-01346  3R01RR014166-16S1                67.0  3R01RR021746-03S1   \n",
       "5    U01DK57304  3U01DK057149-11S1                80.0  3R01DK057030-08S1   \n",
       "6   U01DK057298  3U01DK057219-12S1                91.0  3U01DK061728-09S1   \n",
       "\n",
       "   ARRA_MatchedScore2 ARRA_MatchedGrant3  ARRA_MatchedScore3  \n",
       "0                82.0  3U01DK062503-08S1                82.0  \n",
       "1                82.0  3U01DK057149-11S1                82.0  \n",
       "4                67.0  3R01RR023168-03S1                58.0  \n",
       "5                80.0  3U01DK057131-11S1                80.0  \n",
       "6                82.0  3U01DK072493-04S1                82.0  "
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWAS_Grant.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>State</th>\n",
       "      <th>Organization</th>\n",
       "      <th>City</th>\n",
       "      <th>Award Type</th>\n",
       "      <th>Project Number</th>\n",
       "      <th>DUNS Number</th>\n",
       "      <th>Congressional District</th>\n",
       "      <th>Agency Code</th>\n",
       "      <th>Award Notice Issued</th>\n",
       "      <th>FY Total Cost</th>\n",
       "      <th>...</th>\n",
       "      <th>Activity</th>\n",
       "      <th>IC</th>\n",
       "      <th>Serial Number</th>\n",
       "      <th>Support Year</th>\n",
       "      <th>Suffix</th>\n",
       "      <th>Sub #</th>\n",
       "      <th>Principal Investigator</th>\n",
       "      <th>Admin IC</th>\n",
       "      <th>FY</th>\n",
       "      <th>Project End Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>CA</td>\n",
       "      <td>23ANDME, INC.</td>\n",
       "      <td>MOUNTAIN VIEW</td>\n",
       "      <td>Grant</td>\n",
       "      <td>1R43HG005807-01</td>\n",
       "      <td>780119710</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7529</td>\n",
       "      <td>2010-09-05</td>\n",
       "      <td>189844</td>\n",
       "      <td>...</td>\n",
       "      <td>R43</td>\n",
       "      <td>HG</td>\n",
       "      <td>5807</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MOUNTAIN, JOANNA L.</td>\n",
       "      <td>NHGRI</td>\n",
       "      <td>2010</td>\n",
       "      <td>2012-05-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NC</td>\n",
       "      <td>3-C INSTITUTE FOR SOCIAL DEVELOPMENT</td>\n",
       "      <td>CARY</td>\n",
       "      <td>Grant</td>\n",
       "      <td>1R43RR030780-01</td>\n",
       "      <td>46981549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7529</td>\n",
       "      <td>2010-07-08</td>\n",
       "      <td>199995</td>\n",
       "      <td>...</td>\n",
       "      <td>R43</td>\n",
       "      <td>RR</td>\n",
       "      <td>30780</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DEROSIER, MELISSA E.</td>\n",
       "      <td>NCRR</td>\n",
       "      <td>2010</td>\n",
       "      <td>2011-07-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NC</td>\n",
       "      <td>3-C INSTITUTE FOR SOCIAL DEVELOPMENT</td>\n",
       "      <td>CARY</td>\n",
       "      <td>Grant</td>\n",
       "      <td>3R43MH086983-02S1</td>\n",
       "      <td>46981549</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7529</td>\n",
       "      <td>2010-09-27</td>\n",
       "      <td>240698</td>\n",
       "      <td>...</td>\n",
       "      <td>R43</td>\n",
       "      <td>MH</td>\n",
       "      <td>86983</td>\n",
       "      <td>2</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWICK, DANIELLE CORI</td>\n",
       "      <td>NIMH</td>\n",
       "      <td>2010</td>\n",
       "      <td>2011-09-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>MA</td>\n",
       "      <td>4S3 BIOSCIENCE, INC.</td>\n",
       "      <td>MEDFORD</td>\n",
       "      <td>Grant</td>\n",
       "      <td>1R21NS066269-01</td>\n",
       "      <td>828286703</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7529</td>\n",
       "      <td>2009-07-20</td>\n",
       "      <td>141092</td>\n",
       "      <td>...</td>\n",
       "      <td>R21</td>\n",
       "      <td>NS</td>\n",
       "      <td>66269</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARMSTRONG, DUSTIN</td>\n",
       "      <td>NINDS</td>\n",
       "      <td>2009</td>\n",
       "      <td>2011-07-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>MA</td>\n",
       "      <td>4S3 BIOSCIENCE, INC.</td>\n",
       "      <td>MEDFORD</td>\n",
       "      <td>Grant</td>\n",
       "      <td>5R21NS066269-02</td>\n",
       "      <td>828286703</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7529</td>\n",
       "      <td>2010-06-24</td>\n",
       "      <td>195590</td>\n",
       "      <td>...</td>\n",
       "      <td>R21</td>\n",
       "      <td>NS</td>\n",
       "      <td>66269</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARMSTRONG, DUSTIN</td>\n",
       "      <td>NINDS</td>\n",
       "      <td>2010</td>\n",
       "      <td>2013-07-31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  State                          Organization           City Award Type  \\\n",
       "0    CA                         23ANDME, INC.  MOUNTAIN VIEW      Grant   \n",
       "1    NC  3-C INSTITUTE FOR SOCIAL DEVELOPMENT           CARY      Grant   \n",
       "2    NC  3-C INSTITUTE FOR SOCIAL DEVELOPMENT           CARY      Grant   \n",
       "3    MA                  4S3 BIOSCIENCE, INC.        MEDFORD      Grant   \n",
       "4    MA                  4S3 BIOSCIENCE, INC.        MEDFORD      Grant   \n",
       "\n",
       "      Project Number DUNS Number  Congressional District  Agency Code  \\\n",
       "0    1R43HG005807-01   780119710                    14.0         7529   \n",
       "1    1R43RR030780-01    46981549                     2.0         7529   \n",
       "2  3R43MH086983-02S1    46981549                     2.0         7529   \n",
       "3    1R21NS066269-01   828286703                     7.0         7529   \n",
       "4    5R21NS066269-02   828286703                     7.0         7529   \n",
       "\n",
       "  Award Notice Issued  FY Total Cost  ...  Activity  IC Serial Number  \\\n",
       "0          2010-09-05         189844  ...       R43  HG          5807   \n",
       "1          2010-07-08         199995  ...       R43  RR         30780   \n",
       "2          2010-09-27         240698  ...       R43  MH         86983   \n",
       "3          2009-07-20         141092  ...       R21  NS         66269   \n",
       "4          2010-06-24         195590  ...       R21  NS         66269   \n",
       "\n",
       "  Support Year  Suffix Sub # Principal Investigator  Admin IC    FY  \\\n",
       "0            1     NaN   NaN    MOUNTAIN, JOANNA L.     NHGRI  2010   \n",
       "1            1     NaN   NaN   DEROSIER, MELISSA E.      NCRR  2010   \n",
       "2            2      S1   NaN   SWICK, DANIELLE CORI      NIMH  2010   \n",
       "3            1     NaN   NaN     ARMSTRONG, DUSTIN      NINDS  2009   \n",
       "4            2     NaN   NaN     ARMSTRONG, DUSTIN      NINDS  2010   \n",
       "\n",
       "  Project End Date  \n",
       "0       2012-05-31  \n",
       "1       2011-07-07  \n",
       "2       2011-09-26  \n",
       "3       2011-07-31  \n",
       "4       2013-07-31  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARRA.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CAUTIOUS: Project Number '3R37DK044083-19S1\" is duplicated twice in the ARRA list. It is the only one with duplicates. The info in these two duplicates are the same. So we are safe to drop one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21712, 25)"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARRA = ARRA.drop_duplicates('Project Number')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge the ARRA Grant detailed information to the GWAS Grant list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_ARRA_Grant = pd.merge(GWAS_Grant, ARRA, left_on =\"ARRA_MatchedGrant1\", right_on =\"Project Number\", \n",
    "                           validate = \"m:1\", indicator = \"ARRA_merge\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GrantID</th>\n",
       "      <th>ARRA_MatchedGrant1</th>\n",
       "      <th>ARRA_MatchedScore1</th>\n",
       "      <th>ARRA_MatchedGrant2</th>\n",
       "      <th>ARRA_MatchedScore2</th>\n",
       "      <th>ARRA_MatchedGrant3</th>\n",
       "      <th>ARRA_MatchedScore3</th>\n",
       "      <th>State</th>\n",
       "      <th>Organization</th>\n",
       "      <th>City</th>\n",
       "      <th>...</th>\n",
       "      <th>IC</th>\n",
       "      <th>Serial Number</th>\n",
       "      <th>Support Year</th>\n",
       "      <th>Suffix</th>\n",
       "      <th>Sub #</th>\n",
       "      <th>Principal Investigator</th>\n",
       "      <th>Admin IC</th>\n",
       "      <th>FY</th>\n",
       "      <th>Project End Date</th>\n",
       "      <th>ARRA_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>U01DK057303</td>\n",
       "      <td>3U01DK061730-08S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK061730-09S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK062503-08S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>MD</td>\n",
       "      <td>JOHNS HOPKINS UNIVERSITY</td>\n",
       "      <td>BALTIMORE</td>\n",
       "      <td>...</td>\n",
       "      <td>DK</td>\n",
       "      <td>61730</td>\n",
       "      <td>8</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TONASCIA, JAMES A</td>\n",
       "      <td>NIDDK</td>\n",
       "      <td>2010</td>\n",
       "      <td>2011-01-31</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>U01DK061730</td>\n",
       "      <td>3U01DK061730-08S1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3U01DK061730-09S1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3U01DK061732-08S1</td>\n",
       "      <td>91.0</td>\n",
       "      <td>MD</td>\n",
       "      <td>JOHNS HOPKINS UNIVERSITY</td>\n",
       "      <td>BALTIMORE</td>\n",
       "      <td>...</td>\n",
       "      <td>DK</td>\n",
       "      <td>61730</td>\n",
       "      <td>8</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TONASCIA, JAMES A</td>\n",
       "      <td>NIDDK</td>\n",
       "      <td>2010</td>\n",
       "      <td>2011-01-31</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>U01DK057304</td>\n",
       "      <td>3U01DK061734-09S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK072504-05S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3U01DK057149-11S1</td>\n",
       "      <td>82.0</td>\n",
       "      <td>NY</td>\n",
       "      <td>COLUMBIA UNIVERSITY HEALTH SCIENCES</td>\n",
       "      <td>NEW YORK</td>\n",
       "      <td>...</td>\n",
       "      <td>DK</td>\n",
       "      <td>61734</td>\n",
       "      <td>9</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAVINE, JOEL EDWARD</td>\n",
       "      <td>NIDDK</td>\n",
       "      <td>2010</td>\n",
       "      <td>2012-01-31</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>U01DK061734</td>\n",
       "      <td>3U01DK061734-09S1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3U01DK066174-08S1</td>\n",
       "      <td>91.0</td>\n",
       "      <td>3U01DK066174-08S2</td>\n",
       "      <td>91.0</td>\n",
       "      <td>NY</td>\n",
       "      <td>COLUMBIA UNIVERSITY HEALTH SCIENCES</td>\n",
       "      <td>NEW YORK</td>\n",
       "      <td>...</td>\n",
       "      <td>DK</td>\n",
       "      <td>61734</td>\n",
       "      <td>9</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAVINE, JOEL EDWARD</td>\n",
       "      <td>NIDDK</td>\n",
       "      <td>2010</td>\n",
       "      <td>2012-01-31</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M01-RR-01346</td>\n",
       "      <td>3R01RR014166-16S1</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3R01RR021746-03S1</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3R01RR023168-03S1</td>\n",
       "      <td>58.0</td>\n",
       "      <td>AZ</td>\n",
       "      <td>ARIZONA STATE UNIVERSITY-TEMPE CAMPUS</td>\n",
       "      <td>TEMPE</td>\n",
       "      <td>...</td>\n",
       "      <td>RR</td>\n",
       "      <td>14166</td>\n",
       "      <td>16</td>\n",
       "      <td>S1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SMITH, BRIAN H.</td>\n",
       "      <td>NCRR</td>\n",
       "      <td>2009</td>\n",
       "      <td>2011-09-16</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GrantID ARRA_MatchedGrant1  ARRA_MatchedScore1 ARRA_MatchedGrant2  \\\n",
       "0   U01DK057303  3U01DK061730-08S1                82.0  3U01DK061730-09S1   \n",
       "1   U01DK061730  3U01DK061730-08S1               100.0  3U01DK061730-09S1   \n",
       "2   U01DK057304  3U01DK061734-09S1                82.0  3U01DK072504-05S1   \n",
       "3   U01DK061734  3U01DK061734-09S1               100.0  3U01DK066174-08S1   \n",
       "4  M01-RR-01346  3R01RR014166-16S1                67.0  3R01RR021746-03S1   \n",
       "\n",
       "   ARRA_MatchedScore2 ARRA_MatchedGrant3  ARRA_MatchedScore3 State  \\\n",
       "0                82.0  3U01DK062503-08S1                82.0    MD   \n",
       "1               100.0  3U01DK061732-08S1                91.0    MD   \n",
       "2                82.0  3U01DK057149-11S1                82.0    NY   \n",
       "3                91.0  3U01DK066174-08S2                91.0    NY   \n",
       "4                67.0  3R01RR023168-03S1                58.0    AZ   \n",
       "\n",
       "                            Organization       City  ...  IC Serial Number  \\\n",
       "0               JOHNS HOPKINS UNIVERSITY  BALTIMORE  ...  DK         61730   \n",
       "1               JOHNS HOPKINS UNIVERSITY  BALTIMORE  ...  DK         61730   \n",
       "2    COLUMBIA UNIVERSITY HEALTH SCIENCES   NEW YORK  ...  DK         61734   \n",
       "3    COLUMBIA UNIVERSITY HEALTH SCIENCES   NEW YORK  ...  DK         61734   \n",
       "4  ARIZONA STATE UNIVERSITY-TEMPE CAMPUS      TEMPE  ...  RR         14166   \n",
       "\n",
       "  Support Year  Suffix  Sub # Principal Investigator  Admin IC    FY  \\\n",
       "0            8      S1    NaN      TONASCIA, JAMES A     NIDDK  2010   \n",
       "1            8      S1    NaN      TONASCIA, JAMES A     NIDDK  2010   \n",
       "2            9      S1    NaN    LAVINE, JOEL EDWARD     NIDDK  2010   \n",
       "3            9      S1    NaN    LAVINE, JOEL EDWARD     NIDDK  2010   \n",
       "4           16      S1    NaN        SMITH, BRIAN H.      NCRR  2009   \n",
       "\n",
       "  Project End Date ARRA_merge  \n",
       "0       2011-01-31       both  \n",
       "1       2011-01-31       both  \n",
       "2       2012-01-31       both  \n",
       "3       2012-01-31       both  \n",
       "4       2011-09-16       both  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWAS_ARRA_Grant.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_funding = pd.read_csv(os.path.abspath(os.path.join('__file__',\n",
    "                                                         '../..',\n",
    "                                                         'data',\n",
    "                                                         'PUBMED',\n",
    "                                                         'Pubmed_FunderInfo.csv')),\n",
    "                             header=0, encoding='utf-8',\n",
    "                             index_col=False, sep =',')\n",
    "\n",
    "GWAS_funding.columns = GWAS_funding.columns.str.strip()\n",
    "GWAS_funding['GrantID'] = GWAS_funding['GrantID'].str.replace(' ', '')\n",
    "GWAS_funding['GrantID'] = GWAS_funding['GrantID'].fillna('NOT REPORTED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_ARRA_Grant = pd.merge(GWAS_funding, GWAS_ARRA_Grant, left_on =\"GrantID\", right_on =\"GrantID\", \n",
    "                           validate = \"m:1\", indicator = \"funding_merge\" )\n",
    "\n",
    "GWAS_ARRA_Grant.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_ARRA_Grant['IsARRA'] = (GWAS_ARRA_Grant['ARRA_MatchedScore1'] >= 97)\n",
    "GWAS_ARRA_Grant['IsARRA'].astype(int)\n",
    "\n",
    "GWAS_ARRA_Grant = GWAS_ARRA_Grant.groupby('PUBMEDID')['IsARRA'].sum()\n",
    "\n",
    "GWAS_ARRA_Grant = GWAS_ARRA_Grant.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['rep_Hispanic', 'Funded_AHRQ', 'Funded', 'ini_African_American',\n",
       "       'DATE_x', 'ini_N', 'REPLICATION SAMPLE SIZE', 'ini_NR', 'Multiple_EFO',\n",
       "       'ini_Mixed', 'PUBMEDID_y', 'ini_African', 'ini_European', 'ini_Asian',\n",
       "       'rep_European', 'ini_Hispanic', 'FIRST AUTHOR', 'STUDY',\n",
       "       'GENOTYPING TECHNOLOGY', 'Funded_HHS', 'rep_N', 'INITIAL SAMPLE SIZE',\n",
       "       'DISEASE/TRAIT', 'rep_African', 'CCS_merge', 'DATE ADDED TO CATALOG',\n",
       "       'rep_NR', 'PLATFORM [SNPS PASSING QC]', 'MAPPED_TRAIT', 'rep_Mixed',\n",
       "       'rep_African_American', 'STUDY ACCESSION', 'Funding_merge', 'rep_Asian',\n",
       "       'PUBMEDID_x', 'MAPPED_TRAIT_URI', 'ASSOCIATION COUNT', 'Funded_US',\n",
       "       'DATE_y', 'Funded_NIH', 'EFO', 'LINK', 'JOURNAL', 'str', 'Funded_ARRA',\n",
       "       'CCS', 'Disease', 'N_African', 'N_African_American', 'N_Asian',\n",
       "       'N_European', 'N_Hispanic', 'N_Mixed', 'N_NR', 'N'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestry = pd.merge(Ancestry, GWAS_ARRA_Grant, left_on = 'PUBMEDID_y', right_on = \"PUBMEDID\", \n",
    "                                      how = 'left', validate = 'm:1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ancestry.to_csv(os.path.abspath(os.path.join('__file__',\n",
    "                                        '../..',\n",
    "                                        'data',\n",
    "                                        'Catalogue',\n",
    "                                        'Synthetic',\n",
    "                                        'Ancestral_Study_CCS.csv')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
