{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze the function **load_gwas_cat()** in *LoadData.py*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load the cat_stud data file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>ASSOCIATION COUNT</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2019-08-23</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31434886</td>\n",
       "      <td>Early progression to active tuberculosis is a ...</td>\n",
       "      <td>Early progression to active pulmonary tubercul...</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix [7756401] (imputed)</td>\n",
       "      <td>1</td>\n",
       "      <td>Tuberculosis, susceptibility to Mycobacterium ...</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_3389, http:...</td>\n",
       "      <td>GCST008488</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;90th survival percentile)</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>Affymetrix, Illumina [up to 10786384] (imputed)</td>\n",
       "      <td>2</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008598</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;99th survival percentile)</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>Illumina [up to 9449311] (imputed)</td>\n",
       "      <td>2</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008599</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;90th survival percentile)</td>\n",
       "      <td>11,262 European ancestry cases, 2,178 East Asi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix, Illumina [up to 10786384] (imputed)</td>\n",
       "      <td>8</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008600</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;99th survival percentile)</td>\n",
       "      <td>3,484 European ancestry cases, 2,178 East Asia...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina [up to 9449311] (imputed)</td>\n",
       "      <td>7</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008601</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID FIRST AUTHOR        DATE     JOURNAL  \\\n",
       "0            2019-08-23  31434886        Luo Y  2019-08-21  Nat Commun   \n",
       "1            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "2            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "3            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "4            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/31434886   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Early progression to active tuberculosis is a ...   \n",
       "1  A meta-analysis of genome-wide association stu...   \n",
       "2  A meta-analysis of genome-wide association stu...   \n",
       "3  A meta-analysis of genome-wide association stu...   \n",
       "4  A meta-analysis of genome-wide association stu...   \n",
       "\n",
       "                                       DISEASE/TRAIT  \\\n",
       "0  Early progression to active pulmonary tubercul...   \n",
       "1          Longevity (age >90th survival percentile)   \n",
       "2          Longevity (age >99th survival percentile)   \n",
       "3          Longevity (age >90th survival percentile)   \n",
       "4          Longevity (age >99th survival percentile)   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  3,484 European ancestry cases, 25,483 European...   \n",
       "3  11,262 European ancestry cases, 2,178 East Asi...   \n",
       "4  3,484 European ancestry cases, 2,178 East Asia...   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE  \\\n",
       "0                                                NaN   \n",
       "1  2,557 European ancestry cases, 4,987 European ...   \n",
       "2  1,911 European ancestry cases, 4,987 European ...   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "                        PLATFORM [SNPS PASSING QC]  ASSOCIATION COUNT  \\\n",
       "0                   Affymetrix [7756401] (imputed)                  1   \n",
       "1  Affymetrix, Illumina [up to 10786384] (imputed)                  2   \n",
       "2               Illumina [up to 9449311] (imputed)                  2   \n",
       "3  Affymetrix, Illumina [up to 10786384] (imputed)                  8   \n",
       "4               Illumina [up to 9449311] (imputed)                  7   \n",
       "\n",
       "                                        MAPPED_TRAIT  \\\n",
       "0  Tuberculosis, susceptibility to Mycobacterium ...   \n",
       "1                                          longevity   \n",
       "2                                          longevity   \n",
       "3                                          longevity   \n",
       "4                                          longevity   \n",
       "\n",
       "                                    MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  http://www.orpha.net/ORDO/Orphanet_3389, http:...      GCST008488   \n",
       "1               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008598   \n",
       "2               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008599   \n",
       "3               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008600   \n",
       "4               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008601   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY  \n",
       "0  Genome-wide genotyping array  \n",
       "1  Genome-wide genotyping array  \n",
       "2  Genome-wide genotyping array  \n",
       "3  Genome-wide genotyping array  \n",
       "4  Genome-wide genotyping array  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Stud = pd.read_csv(os.path.abspath(\n",
    "                           os.path.join('__file__',\n",
    "                                        '..',\n",
    "                                        'data',\n",
    "                                        'Catalogue',\n",
    "                                        'Raw',\n",
    "                                        'Cat_Stud.tsv')),\n",
    "                           header=0, sep='\\t', encoding='utf-8',\n",
    "                           index_col=False)\n",
    "Cat_Stud.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>ASSOCIATION COUNT</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2019-08-23</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31434886</td>\n",
       "      <td>Early progression to active tuberculosis is a ...</td>\n",
       "      <td>Early progression to active pulmonary tubercul...</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix [7756401] (imputed)</td>\n",
       "      <td>1</td>\n",
       "      <td>Tuberculosis, susceptibility to Mycobacterium ...</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_3389, http:...</td>\n",
       "      <td>GCST008488</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;90th survival percentile)</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>Affymetrix, Illumina [up to 10786384] (imputed)</td>\n",
       "      <td>2</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008598</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;99th survival percentile)</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>Illumina [up to 9449311] (imputed)</td>\n",
       "      <td>2</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008599</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;90th survival percentile)</td>\n",
       "      <td>11,262 European ancestry cases, 2,178 East Asi...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Affymetrix, Illumina [up to 10786384] (imputed)</td>\n",
       "      <td>8</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008600</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019-09-05</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31413261</td>\n",
       "      <td>A meta-analysis of genome-wide association stu...</td>\n",
       "      <td>Longevity (age &gt;99th survival percentile)</td>\n",
       "      <td>3,484 European ancestry cases, 2,178 East Asia...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>Illumina [up to 9449311] (imputed)</td>\n",
       "      <td>7</td>\n",
       "      <td>longevity</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004300</td>\n",
       "      <td>GCST008601</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID FIRST AUTHOR        DATE     JOURNAL  \\\n",
       "0            2019-08-23  31434886        Luo Y  2019-08-21  Nat Commun   \n",
       "1            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "2            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "3            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "4            2019-09-05  31413261     Deelen J  2019-08-14  Nat Commun   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/31434886   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/31413261   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Early progression to active tuberculosis is a ...   \n",
       "1  A meta-analysis of genome-wide association stu...   \n",
       "2  A meta-analysis of genome-wide association stu...   \n",
       "3  A meta-analysis of genome-wide association stu...   \n",
       "4  A meta-analysis of genome-wide association stu...   \n",
       "\n",
       "                                       DISEASE/TRAIT  \\\n",
       "0  Early progression to active pulmonary tubercul...   \n",
       "1          Longevity (age >90th survival percentile)   \n",
       "2          Longevity (age >99th survival percentile)   \n",
       "3          Longevity (age >90th survival percentile)   \n",
       "4          Longevity (age >99th survival percentile)   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  3,484 European ancestry cases, 25,483 European...   \n",
       "3  11,262 European ancestry cases, 2,178 East Asi...   \n",
       "4  3,484 European ancestry cases, 2,178 East Asia...   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE  \\\n",
       "0                                                N/A   \n",
       "1  2,557 European ancestry cases, 4,987 European ...   \n",
       "2  1,911 European ancestry cases, 4,987 European ...   \n",
       "3                                                N/A   \n",
       "4                                                N/A   \n",
       "\n",
       "                        PLATFORM [SNPS PASSING QC]  ASSOCIATION COUNT  \\\n",
       "0                   Affymetrix [7756401] (imputed)                  1   \n",
       "1  Affymetrix, Illumina [up to 10786384] (imputed)                  2   \n",
       "2               Illumina [up to 9449311] (imputed)                  2   \n",
       "3  Affymetrix, Illumina [up to 10786384] (imputed)                  8   \n",
       "4               Illumina [up to 9449311] (imputed)                  7   \n",
       "\n",
       "                                        MAPPED_TRAIT  \\\n",
       "0  Tuberculosis, susceptibility to Mycobacterium ...   \n",
       "1                                          longevity   \n",
       "2                                          longevity   \n",
       "3                                          longevity   \n",
       "4                                          longevity   \n",
       "\n",
       "                                    MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  http://www.orpha.net/ORDO/Orphanet_3389, http:...      GCST008488   \n",
       "1               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008598   \n",
       "2               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008599   \n",
       "3               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008600   \n",
       "4               http://www.ebi.ac.uk/efo/EFO_0004300      GCST008601   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY  \n",
       "0  Genome-wide genotyping array  \n",
       "1  Genome-wide genotyping array  \n",
       "2  Genome-wide genotyping array  \n",
       "3  Genome-wide genotyping array  \n",
       "4  Genome-wide genotyping array  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fill in the missing value with NA\n",
    "Cat_Stud.fillna('N/A', inplace=True)\n",
    "Cat_Stud.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load the ancenstry data *Cat_Anc*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST008488</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>3980.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peruvian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>36745.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>France, U.S., Netherlands, Denmark, Finland, I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>7544.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>28967.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S., France, Netherlands, Denmark, Finland, U...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>6898.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCCESSION  PUBMEDID FIRST AUTHOR        DATE  \\\n",
       "0       GCST008488  31434886        Luo Y  2019-08-21   \n",
       "1       GCST008598  31413261     Deelen J  2019-08-14   \n",
       "2       GCST008598  31413261     Deelen J  2019-08-14   \n",
       "3       GCST008599  31413261     Deelen J  2019-08-14   \n",
       "4       GCST008599  31413261     Deelen J  2019-08-14   \n",
       "\n",
       "                          INITIAL SAMPLE DESCRIPTION  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  11,262 European ancestry cases, 25,483 Europea...   \n",
       "3  3,484 European ancestry cases, 25,483 European...   \n",
       "4  3,484 European ancestry cases, 25,483 European...   \n",
       "\n",
       "                      REPLICATION SAMPLE DESCRIPTION        STAGE        N  \\\n",
       "0                                                NaN      initial   3980.0   \n",
       "1  2,557 European ancestry cases, 4,987 European ...      initial  36745.0   \n",
       "2  2,557 European ancestry cases, 4,987 European ...  replication   7544.0   \n",
       "3  1,911 European ancestry cases, 4,987 European ...      initial  28967.0   \n",
       "4  1,911 European ancestry cases, 4,987 European ...  replication   6898.0   \n",
       "\n",
       "  BROAD ANCESTRAL COUNTRY OF ORIGIN  \\\n",
       "0              NR                NR   \n",
       "1        European                NR   \n",
       "2        European                NR   \n",
       "3        European                NR   \n",
       "4        European                NR   \n",
       "\n",
       "                              COUNTRY OF RECRUITMENT  \\\n",
       "0                                               Peru   \n",
       "1  France, U.S., Netherlands, Denmark, Finland, I...   \n",
       "2                                   Germany, Denmark   \n",
       "3  U.S., France, Netherlands, Denmark, Finland, U...   \n",
       "4                                   Germany, Denmark   \n",
       "\n",
       "  ADDITONAL ANCESTRY DESCRIPTION  \n",
       "0                       Peruvian  \n",
       "1                            NaN  \n",
       "2                            NaN  \n",
       "3                            NaN  \n",
       "4                            NaN  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Anc = pd.read_csv(os.path.abspath(\n",
    "                          os.path.join('__file__', '..',\n",
    "                                       'data',\n",
    "                                       'Catalogue',\n",
    "                                       'Raw',\n",
    "                                       'Cat_Anc.tsv')),\n",
    "                          header=0, sep='\\t', encoding='utf-8',\n",
    "                          index_col=False)\n",
    "\n",
    "Cat_Anc.rename(columns={'BROAD ANCESTRAL CATEGORY': 'BROAD ANCESTRAL',\n",
    "                            'NUMBER OF INDIVDUALS': 'N'}, inplace=True)\n",
    "Cat_Anc.head()\n",
    "# if an article contains more than one ancestry, then the data will breakdown the ancenstries and put them into different rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST008488</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>3980.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peruvian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>36745.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>France, U.S., Netherlands, Denmark, Finland, I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>7544.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>28967.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S., France, Netherlands, Denmark, Finland, U...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>6898.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCCESSION  PUBMEDID FIRST AUTHOR        DATE  \\\n",
       "0       GCST008488  31434886        Luo Y  2019-08-21   \n",
       "1       GCST008598  31413261     Deelen J  2019-08-14   \n",
       "2       GCST008598  31413261     Deelen J  2019-08-14   \n",
       "3       GCST008599  31413261     Deelen J  2019-08-14   \n",
       "4       GCST008599  31413261     Deelen J  2019-08-14   \n",
       "\n",
       "                          INITIAL SAMPLE DESCRIPTION  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  11,262 European ancestry cases, 25,483 Europea...   \n",
       "3  3,484 European ancestry cases, 25,483 European...   \n",
       "4  3,484 European ancestry cases, 25,483 European...   \n",
       "\n",
       "                      REPLICATION SAMPLE DESCRIPTION        STAGE        N  \\\n",
       "0                                                NaN      initial   3980.0   \n",
       "1  2,557 European ancestry cases, 4,987 European ...      initial  36745.0   \n",
       "2  2,557 European ancestry cases, 4,987 European ...  replication   7544.0   \n",
       "3  1,911 European ancestry cases, 4,987 European ...      initial  28967.0   \n",
       "4  1,911 European ancestry cases, 4,987 European ...  replication   6898.0   \n",
       "\n",
       "  BROAD ANCESTRAL COUNTRY OF ORIGIN  \\\n",
       "0              NR                NR   \n",
       "1        European                NR   \n",
       "2        European                NR   \n",
       "3        European                NR   \n",
       "4        European                NR   \n",
       "\n",
       "                              COUNTRY OF RECRUITMENT  \\\n",
       "0                                               Peru   \n",
       "1  France, U.S., Netherlands, Denmark, Finland, I...   \n",
       "2                                   Germany, Denmark   \n",
       "3  U.S., France, Netherlands, Denmark, Finland, U...   \n",
       "4                                   Germany, Denmark   \n",
       "\n",
       "  ADDITONAL ANCESTRY DESCRIPTION  \n",
       "0                       Peruvian  \n",
       "1                            NaN  \n",
       "2                            NaN  \n",
       "3                            NaN  \n",
       "4                            NaN  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop those with null ancenstry articles\n",
    "Cat_Anc = Cat_Anc[~Cat_Anc['BROAD ANCESTRAL'].isnull()]\n",
    "Cat_Anc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST008488</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>3980.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peruvian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>36745.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>France, U.S., Netherlands, Denmark, Finland, I...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>7544.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>28967.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S., France, Netherlands, Denmark, Finland, U...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>6898.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID FIRST AUTHOR        DATE  \\\n",
       "0      GCST008488  31434886        Luo Y  2019-08-21   \n",
       "1      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "2      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "3      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "4      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "\n",
       "                          INITIAL SAMPLE DESCRIPTION  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  11,262 European ancestry cases, 25,483 Europea...   \n",
       "3  3,484 European ancestry cases, 25,483 European...   \n",
       "4  3,484 European ancestry cases, 25,483 European...   \n",
       "\n",
       "                      REPLICATION SAMPLE DESCRIPTION        STAGE        N  \\\n",
       "0                                                NaN      initial   3980.0   \n",
       "1  2,557 European ancestry cases, 4,987 European ...      initial  36745.0   \n",
       "2  2,557 European ancestry cases, 4,987 European ...  replication   7544.0   \n",
       "3  1,911 European ancestry cases, 4,987 European ...      initial  28967.0   \n",
       "4  1,911 European ancestry cases, 4,987 European ...  replication   6898.0   \n",
       "\n",
       "  BROAD ANCESTRAL COUNTRY OF ORIGIN  \\\n",
       "0              NR                NR   \n",
       "1        European                NR   \n",
       "2        European                NR   \n",
       "3        European                NR   \n",
       "4        European                NR   \n",
       "\n",
       "                              COUNTRY OF RECRUITMENT  \\\n",
       "0                                               Peru   \n",
       "1  France, U.S., Netherlands, Denmark, Finland, I...   \n",
       "2                                   Germany, Denmark   \n",
       "3  U.S., France, Netherlands, Denmark, Finland, U...   \n",
       "4                                   Germany, Denmark   \n",
       "\n",
       "  ADDITONAL ANCESTRY DESCRIPTION  \n",
       "0                       Peruvian  \n",
       "1                            NaN  \n",
       "2                            NaN  \n",
       "3                            NaN  \n",
       "4                            NaN  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Replace occurrences of pattern in the the column names with some other string. Here is used to correctify the spelling\n",
    "Cat_Anc.columns = Cat_Anc.columns.str.replace('ACCCESSION', 'ACCESSION')\n",
    "Cat_Anc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collapse by article, and obtain the sample size information for each article.\n",
    "Remark: Currently this sum of sample size for each article is the total of replication sample and initial sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST000001</td>\n",
       "      <td>146.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>1550.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST000003</td>\n",
       "      <td>12316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST000004</td>\n",
       "      <td>5051.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST000005</td>\n",
       "      <td>537.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION        N\n",
       "0      GCST000001    146.0\n",
       "1      GCST000002   1550.0\n",
       "2      GCST000003  12316.0\n",
       "3      GCST000004   5051.0\n",
       "4      GCST000005    537.0"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Anc_byN = Cat_Anc[['STUDY ACCESSION', 'N',\n",
    "                         'DATE']].groupby(by='STUDY ACCESSION').sum()\n",
    "Cat_Anc_byN = Cat_Anc_byN.reset_index()\n",
    "Cat_Anc_byN.head()\n",
    "\n",
    "# get the initial sample size fo each article\n",
    "# Cat_Anc_byN = Cat_Anc[Cat_Anc[\"STAGE\"] == \"initial\"][[\"STUDY ACCESSION\", \"N\", \"DATE\"]].groupby(\"STUDY ACCESSION\").sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>N</th>\n",
       "      <th>DATE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST000001</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2005-03-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>1550.0</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST000003</td>\n",
       "      <td>12316.0</td>\n",
       "      <td>2006-04-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST000004</td>\n",
       "      <td>5051.0</td>\n",
       "      <td>2006-04-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST000005</td>\n",
       "      <td>537.0</td>\n",
       "      <td>2006-09-28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION        N        DATE\n",
       "0      GCST000001    146.0  2005-03-10\n",
       "1      GCST000002   1550.0  2005-09-09\n",
       "2      GCST000003  12316.0  2006-04-14\n",
       "3      GCST000004   5051.0  2006-04-30\n",
       "4      GCST000005    537.0  2006-09-28"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge the ancestral summary back with the gwas article info.\n",
    "Cat_Anc_byN = pd.merge(Cat_Anc_byN, Cat_Stud[[\n",
    "        'STUDY ACCESSION', 'DATE']], how='left', on='STUDY ACCESSION')\n",
    "Cat_Anc_byN.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Begin to clean the ancestral categories based on a per-specified ancestral cleaner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load in the ancestral cleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>Broader</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>African</td>\n",
       "      <td>African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>African American or Afro-Caribbean, African un...</td>\n",
       "      <td>Other/Mixed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>African American or Afro-Caribbean, African un...</td>\n",
       "      <td>Other/Mixed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>African American or Afro-Caribbean, Asian unsp...</td>\n",
       "      <td>Other/Mixed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     BROAD ANCESTRAL  \\\n",
       "0                                            African   \n",
       "1                 African American or Afro-Caribbean   \n",
       "2  African American or Afro-Caribbean, African un...   \n",
       "3  African American or Afro-Caribbean, African un...   \n",
       "4  African American or Afro-Caribbean, Asian unsp...   \n",
       "\n",
       "                              Broader  \n",
       "0                             African  \n",
       "1  African American or Afro-Caribbean  \n",
       "2                         Other/Mixed  \n",
       "3                         Other/Mixed  \n",
       "4                         Other/Mixed  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaner_broad = pd.read_csv(os.path.abspath(\n",
    "        os.path.join('__file__',\n",
    "                     '..',\n",
    "                     'data',\n",
    "                     'Support',\n",
    "                     'dict_replacer_broad.tsv')),\n",
    "        sep='\\t', header=0, index_col=False)\n",
    "cleaner_broad.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### merge the **ancenstral cleaner** to the ancenstry data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST008488</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>3980.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peruvian</td>\n",
       "      <td>In Part Not Recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>36745.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>France, U.S., Netherlands, Denmark, Finland, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>7544.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>28967.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S., France, Netherlands, Denmark, Finland, U...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>6898.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID FIRST AUTHOR        DATE  \\\n",
       "0      GCST008488  31434886        Luo Y  2019-08-21   \n",
       "1      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "2      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "3      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "4      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "\n",
       "                          INITIAL SAMPLE DESCRIPTION  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  11,262 European ancestry cases, 25,483 Europea...   \n",
       "3  3,484 European ancestry cases, 25,483 European...   \n",
       "4  3,484 European ancestry cases, 25,483 European...   \n",
       "\n",
       "                      REPLICATION SAMPLE DESCRIPTION        STAGE        N  \\\n",
       "0                                                NaN      initial   3980.0   \n",
       "1  2,557 European ancestry cases, 4,987 European ...      initial  36745.0   \n",
       "2  2,557 European ancestry cases, 4,987 European ...  replication   7544.0   \n",
       "3  1,911 European ancestry cases, 4,987 European ...      initial  28967.0   \n",
       "4  1,911 European ancestry cases, 4,987 European ...  replication   6898.0   \n",
       "\n",
       "  BROAD ANCESTRAL COUNTRY OF ORIGIN  \\\n",
       "0              NR                NR   \n",
       "1        European                NR   \n",
       "2        European                NR   \n",
       "3        European                NR   \n",
       "4        European                NR   \n",
       "\n",
       "                              COUNTRY OF RECRUITMENT  \\\n",
       "0                                               Peru   \n",
       "1  France, U.S., Netherlands, Denmark, Finland, I...   \n",
       "2                                   Germany, Denmark   \n",
       "3  U.S., France, Netherlands, Denmark, Finland, U...   \n",
       "4                                   Germany, Denmark   \n",
       "\n",
       "  ADDITONAL ANCESTRY DESCRIPTION               Broader  \n",
       "0                       Peruvian  In Part Not Recorded  \n",
       "1                            NaN              European  \n",
       "2                            NaN              European  \n",
       "3                            NaN              European  \n",
       "4                            NaN              European  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Anc = pd.merge(Cat_Anc, cleaner_broad, how='left',\n",
    "                       on='BROAD ANCESTRAL')\n",
    "Cat_Anc.head()\n",
    "# REMARK: the column \"Broader\" is the merged info from the ancestral cleaner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "      <th>Dates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST008488</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>3980.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peruvian</td>\n",
       "      <td>In Part Not Recorded</td>\n",
       "      <td>2019-08-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>36745.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>France, U.S., Netherlands, Denmark, Finland, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>7544.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>28967.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S., France, Netherlands, Denmark, Finland, U...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>6898.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID FIRST AUTHOR        DATE  \\\n",
       "0      GCST008488  31434886        Luo Y  2019-08-21   \n",
       "1      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "2      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "3      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "4      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "\n",
       "                          INITIAL SAMPLE DESCRIPTION  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  11,262 European ancestry cases, 25,483 Europea...   \n",
       "3  3,484 European ancestry cases, 25,483 European...   \n",
       "4  3,484 European ancestry cases, 25,483 European...   \n",
       "\n",
       "                      REPLICATION SAMPLE DESCRIPTION        STAGE        N  \\\n",
       "0                                                NaN      initial   3980.0   \n",
       "1  2,557 European ancestry cases, 4,987 European ...      initial  36745.0   \n",
       "2  2,557 European ancestry cases, 4,987 European ...  replication   7544.0   \n",
       "3  1,911 European ancestry cases, 4,987 European ...      initial  28967.0   \n",
       "4  1,911 European ancestry cases, 4,987 European ...  replication   6898.0   \n",
       "\n",
       "  BROAD ANCESTRAL COUNTRY OF ORIGIN  \\\n",
       "0              NR                NR   \n",
       "1        European                NR   \n",
       "2        European                NR   \n",
       "3        European                NR   \n",
       "4        European                NR   \n",
       "\n",
       "                              COUNTRY OF RECRUITMENT  \\\n",
       "0                                               Peru   \n",
       "1  France, U.S., Netherlands, Denmark, Finland, I...   \n",
       "2                                   Germany, Denmark   \n",
       "3  U.S., France, Netherlands, Denmark, Finland, U...   \n",
       "4                                   Germany, Denmark   \n",
       "\n",
       "  ADDITONAL ANCESTRY DESCRIPTION               Broader      Dates  \n",
       "0                       Peruvian  In Part Not Recorded 2019-08-21  \n",
       "1                            NaN              European 2019-08-14  \n",
       "2                            NaN              European 2019-08-14  \n",
       "3                            NaN              European 2019-08-14  \n",
       "4                            NaN              European 2019-08-14  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Anc['Dates'] = [pd.to_datetime(d) for d in Cat_Anc['DATE']]\n",
    "# creates a new date column \"Dates\", but why?\n",
    "#print(type(Cat_Anc.loc[0,'DATE']))  The column \"DATE\" 's element is a str\n",
    "#print(type(Cat_Anc.loc[0, \"Dates\"])) The column \"Dates\"'s element is a time type\n",
    "Cat_Anc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'numpy.float64'>\n",
      "<class 'numpy.float64'>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "      <th>Dates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST008488</td>\n",
       "      <td>31434886</td>\n",
       "      <td>Luo Y</td>\n",
       "      <td>2019-08-21</td>\n",
       "      <td>2,160 Peruvian cases, 1,820 Peruvian controls</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>3980.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peruvian</td>\n",
       "      <td>In Part Not Recorded</td>\n",
       "      <td>2019-08-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>36745.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>France, U.S., Netherlands, Denmark, Finland, I...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST008598</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>11,262 European ancestry cases, 25,483 Europea...</td>\n",
       "      <td>2,557 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>7544.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>initial</td>\n",
       "      <td>28967.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S., France, Netherlands, Denmark, Finland, U...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST008599</td>\n",
       "      <td>31413261</td>\n",
       "      <td>Deelen J</td>\n",
       "      <td>2019-08-14</td>\n",
       "      <td>3,484 European ancestry cases, 25,483 European...</td>\n",
       "      <td>1,911 European ancestry cases, 4,987 European ...</td>\n",
       "      <td>replication</td>\n",
       "      <td>6898.0</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>Germany, Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2019-08-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  STUDY ACCESSION  PUBMEDID FIRST AUTHOR        DATE  \\\n",
       "0      GCST008488  31434886        Luo Y  2019-08-21   \n",
       "1      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "2      GCST008598  31413261     Deelen J  2019-08-14   \n",
       "3      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "4      GCST008599  31413261     Deelen J  2019-08-14   \n",
       "\n",
       "                          INITIAL SAMPLE DESCRIPTION  \\\n",
       "0      2,160 Peruvian cases, 1,820 Peruvian controls   \n",
       "1  11,262 European ancestry cases, 25,483 Europea...   \n",
       "2  11,262 European ancestry cases, 25,483 Europea...   \n",
       "3  3,484 European ancestry cases, 25,483 European...   \n",
       "4  3,484 European ancestry cases, 25,483 European...   \n",
       "\n",
       "                      REPLICATION SAMPLE DESCRIPTION        STAGE        N  \\\n",
       "0                                                NaN      initial   3980.0   \n",
       "1  2,557 European ancestry cases, 4,987 European ...      initial  36745.0   \n",
       "2  2,557 European ancestry cases, 4,987 European ...  replication   7544.0   \n",
       "3  1,911 European ancestry cases, 4,987 European ...      initial  28967.0   \n",
       "4  1,911 European ancestry cases, 4,987 European ...  replication   6898.0   \n",
       "\n",
       "  BROAD ANCESTRAL COUNTRY OF ORIGIN  \\\n",
       "0              NR                NR   \n",
       "1        European                NR   \n",
       "2        European                NR   \n",
       "3        European                NR   \n",
       "4        European                NR   \n",
       "\n",
       "                              COUNTRY OF RECRUITMENT  \\\n",
       "0                                               Peru   \n",
       "1  France, U.S., Netherlands, Denmark, Finland, I...   \n",
       "2                                   Germany, Denmark   \n",
       "3  U.S., France, Netherlands, Denmark, Finland, U...   \n",
       "4                                   Germany, Denmark   \n",
       "\n",
       "  ADDITONAL ANCESTRY DESCRIPTION               Broader      Dates  \n",
       "0                       Peruvian  In Part Not Recorded 2019-08-21  \n",
       "1                            NaN              European 2019-08-14  \n",
       "2                            NaN              European 2019-08-14  \n",
       "3                            NaN              European 2019-08-14  \n",
       "4                            NaN              European 2019-08-14  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(type(Cat_Anc.loc[0,\"N\"]))\n",
    "Cat_Anc['N'] = pd.to_numeric(Cat_Anc['N'], errors='coerce')\n",
    "Cat_Anc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "      <th>Dates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14706</th>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>Klein RJ</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>146</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-03-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14701</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>141</td>\n",
       "      <td>Other</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other/Mixed</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14703</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>replication</td>\n",
       "      <td>541</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14702</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>1</td>\n",
       "      <td>Asian unspecified</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Asian</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14700</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>744</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      STUDY ACCESSION  PUBMEDID   FIRST AUTHOR        DATE  \\\n",
       "14706      GCST000001  15761122       Klein RJ  2005-03-10   \n",
       "14701      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "14703      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "14702      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "14700      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "\n",
       "                              INITIAL SAMPLE DESCRIPTION  \\\n",
       "14706  96 European ancestry cases, 50 European ancest...   \n",
       "14701  381 European ancestry cases, 363 European ance...   \n",
       "14703  381 European ancestry cases, 363 European ance...   \n",
       "14702  381 European ancestry cases, 363 European ance...   \n",
       "14700  381 European ancestry cases, 363 European ance...   \n",
       "\n",
       "                          REPLICATION SAMPLE DESCRIPTION        STAGE    N  \\\n",
       "14706                                                NaN      initial  146   \n",
       "14701  269 European ancestry cases, 272 European ance...      initial  141   \n",
       "14703  269 European ancestry cases, 272 European ance...  replication  541   \n",
       "14702  269 European ancestry cases, 272 European ance...      initial    1   \n",
       "14700  269 European ancestry cases, 272 European ance...      initial  744   \n",
       "\n",
       "         BROAD ANCESTRAL COUNTRY OF ORIGIN COUNTRY OF RECRUITMENT  \\\n",
       "14706           European                NR                     NR   \n",
       "14701              Other                NR                   U.S.   \n",
       "14703           European                NR                   U.S.   \n",
       "14702  Asian unspecified                NR                   U.S.   \n",
       "14700           European                NR                   U.S.   \n",
       "\n",
       "      ADDITONAL ANCESTRY DESCRIPTION      Broader      Dates  \n",
       "14706                            NaN     European 2005-03-10  \n",
       "14701                            NaN  Other/Mixed 2005-09-09  \n",
       "14703                            NaN     European 2005-09-09  \n",
       "14702                            NaN        Asian 2005-09-09  \n",
       "14700                            NaN     European 2005-09-09  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## keep the non-missing sample size articiles\n",
    "Cat_Anc = Cat_Anc[Cat_Anc['N'].notnull()]\n",
    "\n",
    "## cast the sample size column into integer, previous is float\n",
    "Cat_Anc['N'] = Cat_Anc['N'].astype(int)\n",
    "\n",
    "## sort the data by dates, from the earliest to latest\n",
    "Cat_Anc = Cat_Anc.sort_values(by='Dates', ascending = True)\n",
    "\n",
    "Cat_Anc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>INITIAL SAMPLE DESCRIPTION</th>\n",
       "      <th>REPLICATION SAMPLE DESCRIPTION</th>\n",
       "      <th>STAGE</th>\n",
       "      <th>N</th>\n",
       "      <th>BROAD ANCESTRAL</th>\n",
       "      <th>COUNTRY OF ORIGIN</th>\n",
       "      <th>COUNTRY OF RECRUITMENT</th>\n",
       "      <th>ADDITONAL ANCESTRY DESCRIPTION</th>\n",
       "      <th>Broader</th>\n",
       "      <th>Dates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14706</th>\n",
       "      <td>GCST000001</td>\n",
       "      <td>15761122</td>\n",
       "      <td>Klein RJ</td>\n",
       "      <td>2005-03-10</td>\n",
       "      <td>96 European ancestry cases, 50 European ancest...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>initial</td>\n",
       "      <td>146</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-03-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14701</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>141</td>\n",
       "      <td>Other</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other/Mixed</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14703</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>replication</td>\n",
       "      <td>541</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14702</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>1</td>\n",
       "      <td>Asian unspecified</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Asian</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14700</th>\n",
       "      <td>GCST000002</td>\n",
       "      <td>16252231</td>\n",
       "      <td>Maraganore DM</td>\n",
       "      <td>2005-09-09</td>\n",
       "      <td>381 European ancestry cases, 363 European ance...</td>\n",
       "      <td>269 European ancestry cases, 272 European ance...</td>\n",
       "      <td>initial</td>\n",
       "      <td>744</td>\n",
       "      <td>European</td>\n",
       "      <td>NR</td>\n",
       "      <td>U.S.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "      <td>2005-09-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      STUDY ACCESSION  PUBMEDID   FIRST AUTHOR        DATE  \\\n",
       "14706      GCST000001  15761122       Klein RJ  2005-03-10   \n",
       "14701      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "14703      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "14702      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "14700      GCST000002  16252231  Maraganore DM  2005-09-09   \n",
       "\n",
       "                              INITIAL SAMPLE DESCRIPTION  \\\n",
       "14706  96 European ancestry cases, 50 European ancest...   \n",
       "14701  381 European ancestry cases, 363 European ance...   \n",
       "14703  381 European ancestry cases, 363 European ance...   \n",
       "14702  381 European ancestry cases, 363 European ance...   \n",
       "14700  381 European ancestry cases, 363 European ance...   \n",
       "\n",
       "                          REPLICATION SAMPLE DESCRIPTION        STAGE    N  \\\n",
       "14706                                                NaN      initial  146   \n",
       "14701  269 European ancestry cases, 272 European ance...      initial  141   \n",
       "14703  269 European ancestry cases, 272 European ance...  replication  541   \n",
       "14702  269 European ancestry cases, 272 European ance...      initial    1   \n",
       "14700  269 European ancestry cases, 272 European ance...      initial  744   \n",
       "\n",
       "         BROAD ANCESTRAL COUNTRY OF ORIGIN COUNTRY OF RECRUITMENT  \\\n",
       "14706           European                NR                     NR   \n",
       "14701              Other                NR                   U.S.   \n",
       "14703           European                NR                   U.S.   \n",
       "14702  Asian unspecified                NR                   U.S.   \n",
       "14700           European                NR                   U.S.   \n",
       "\n",
       "      ADDITONAL ANCESTRY DESCRIPTION      Broader      Dates  \n",
       "14706                            NaN     European 2005-03-10  \n",
       "14701                            NaN  Other/Mixed 2005-09-09  \n",
       "14703                            NaN     European 2005-09-09  \n",
       "14702                            NaN        Asian 2005-09-09  \n",
       "14700                            NaN     European 2005-09-09  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Anc['Broader']\n",
    "## Hard coding: replace a typical ancestry category, there pandas.series.str.replace is used.\n",
    "Cat_Anc['Broader'] = Cat_Anc['Broader'].str.replace(\n",
    "        'African American or Afro-Caribbean', 'African Am./Caribbean')\n",
    "\n",
    "Cat_Anc['Broader'] = Cat_Anc['Broader'].str.replace(\n",
    "        'Hispanic or Latin American', 'Hispanic/Latin American')\n",
    "\n",
    "Cat_Anc = Cat_Anc.sort_values(by=[\"PUBMEDID\", 'Dates'], ascending = True)\n",
    "Cat_Anc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the full GWAS article info dataset, which includes all the information, such as p-value, SNP, for each GWAS article."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>INTERGENIC</th>\n",
       "      <th>RISK ALLELE FREQUENCY</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2015-05-20</td>\n",
       "      <td>21801394</td>\n",
       "      <td>Genin E</td>\n",
       "      <td>2011-07-29</td>\n",
       "      <td>Orphanet J Rare Dis</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21801394</td>\n",
       "      <td>Genome-wide association study of Stevens-Johns...</td>\n",
       "      <td>Drug-induced Stevens-Johnson syndrome or toxic...</td>\n",
       "      <td>424 European ancestry cases, 1,881 European an...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.15</td>\n",
       "      <td>2E-9</td>\n",
       "      <td>8.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.73</td>\n",
       "      <td>[1.44-2.08]</td>\n",
       "      <td>Illumina [268914]</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>21743467</td>\n",
       "      <td>Kote-Jarai Z</td>\n",
       "      <td>2011-07-10</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21743467</td>\n",
       "      <td>Seven prostate cancer susceptibility loci iden...</td>\n",
       "      <td>Prostate cancer</td>\n",
       "      <td>6,621 European ancestry cases, 6,939 European ...</td>\n",
       "      <td>22,957 European ancestry cases, 23,234 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>3E-15</td>\n",
       "      <td>14.522879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.09</td>\n",
       "      <td>[1.06-1.12]</td>\n",
       "      <td>NR [2600000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>21743467</td>\n",
       "      <td>Kote-Jarai Z</td>\n",
       "      <td>2011-07-10</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21743467</td>\n",
       "      <td>Seven prostate cancer susceptibility loci iden...</td>\n",
       "      <td>Prostate cancer</td>\n",
       "      <td>6,621 European ancestry cases, 6,939 European ...</td>\n",
       "      <td>22,957 European ancestry cases, 23,234 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.22</td>\n",
       "      <td>3E-9</td>\n",
       "      <td>8.522879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.06</td>\n",
       "      <td>[1.02-1.09]</td>\n",
       "      <td>NR [2600000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>21743467</td>\n",
       "      <td>Kote-Jarai Z</td>\n",
       "      <td>2011-07-10</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21743467</td>\n",
       "      <td>Seven prostate cancer susceptibility loci iden...</td>\n",
       "      <td>Prostate cancer</td>\n",
       "      <td>6,621 European ancestry cases, 6,939 European ...</td>\n",
       "      <td>22,957 European ancestry cases, 23,234 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.45</td>\n",
       "      <td>2E-8</td>\n",
       "      <td>7.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.04</td>\n",
       "      <td>[1.01-1.07]</td>\n",
       "      <td>NR [2600000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>21743467</td>\n",
       "      <td>Kote-Jarai Z</td>\n",
       "      <td>2011-07-10</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21743467</td>\n",
       "      <td>Seven prostate cancer susceptibility loci iden...</td>\n",
       "      <td>Prostate cancer</td>\n",
       "      <td>6,621 European ancestry cases, 6,939 European ...</td>\n",
       "      <td>22,957 European ancestry cases, 23,234 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.52</td>\n",
       "      <td>7E-22</td>\n",
       "      <td>21.154902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.11</td>\n",
       "      <td>[1.08-1.14]</td>\n",
       "      <td>NR [2600000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID  FIRST AUTHOR        DATE  \\\n",
       "0            2015-05-20  21801394       Genin E  2011-07-29   \n",
       "1            2011-08-22  21743467  Kote-Jarai Z  2011-07-10   \n",
       "2            2011-08-22  21743467  Kote-Jarai Z  2011-07-10   \n",
       "3            2011-08-22  21743467  Kote-Jarai Z  2011-07-10   \n",
       "4            2011-08-22  21743467  Kote-Jarai Z  2011-07-10   \n",
       "\n",
       "               JOURNAL                                  LINK  \\\n",
       "0  Orphanet J Rare Dis  www.ncbi.nlm.nih.gov/pubmed/21801394   \n",
       "1            Nat Genet  www.ncbi.nlm.nih.gov/pubmed/21743467   \n",
       "2            Nat Genet  www.ncbi.nlm.nih.gov/pubmed/21743467   \n",
       "3            Nat Genet  www.ncbi.nlm.nih.gov/pubmed/21743467   \n",
       "4            Nat Genet  www.ncbi.nlm.nih.gov/pubmed/21743467   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Genome-wide association study of Stevens-Johns...   \n",
       "1  Seven prostate cancer susceptibility loci iden...   \n",
       "2  Seven prostate cancer susceptibility loci iden...   \n",
       "3  Seven prostate cancer susceptibility loci iden...   \n",
       "4  Seven prostate cancer susceptibility loci iden...   \n",
       "\n",
       "                                       DISEASE/TRAIT  \\\n",
       "0  Drug-induced Stevens-Johnson syndrome or toxic...   \n",
       "1                                    Prostate cancer   \n",
       "2                                    Prostate cancer   \n",
       "3                                    Prostate cancer   \n",
       "4                                    Prostate cancer   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0  424 European ancestry cases, 1,881 European an...   \n",
       "1  6,621 European ancestry cases, 6,939 European ...   \n",
       "2  6,621 European ancestry cases, 6,939 European ...   \n",
       "3  6,621 European ancestry cases, 6,939 European ...   \n",
       "4  6,621 European ancestry cases, 6,939 European ...   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE ...              CONTEXT  \\\n",
       "0                                                NaN ...       intron_variant   \n",
       "1  22,957 European ancestry cases, 23,234 Europea... ...       intron_variant   \n",
       "2  22,957 European ancestry cases, 23,234 Europea... ...   intergenic_variant   \n",
       "3  22,957 European ancestry cases, 23,234 Europea... ...       intron_variant   \n",
       "4  22,957 European ancestry cases, 23,234 Europea... ...       intron_variant   \n",
       "\n",
       "  INTERGENIC RISK ALLELE FREQUENCY P-VALUE PVALUE_MLOG P-VALUE (TEXT)  \\\n",
       "0        0.0                  0.15    2E-9    8.698970            NaN   \n",
       "1        0.0                  0.59   3E-15   14.522879            NaN   \n",
       "2        1.0                  0.22    3E-9    8.522879            NaN   \n",
       "3        0.0                  0.45    2E-8    7.698970            NaN   \n",
       "4        0.0                  0.52   7E-22   21.154902            NaN   \n",
       "\n",
       "  OR or BETA 95% CI (TEXT)  PLATFORM [SNPS PASSING QC]  CNV  \n",
       "0       1.73   [1.44-2.08]           Illumina [268914]    N  \n",
       "1       1.09   [1.06-1.12]      NR [2600000] (imputed)    N  \n",
       "2       1.06   [1.02-1.09]      NR [2600000] (imputed)    N  \n",
       "3       1.04   [1.01-1.07]      NR [2600000] (imputed)    N  \n",
       "4       1.11   [1.08-1.14]      NR [2600000] (imputed)    N  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cat_Full = pd.read_csv(os.path.abspath(os.path.join('__file__',\n",
    "                                                        '..',\n",
    "                                                        'data',\n",
    "                                                        'Catalogue',\n",
    "                                                        'Raw',\n",
    "                                                        'Cat_Full.tsv')),\n",
    "                           header=0, sep='\\t', encoding='utf-8',\n",
    "                           index_col=False, low_memory=False)\n",
    "Cat_Full.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the merged ancestry data set into csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "Cat_Anc.to_csv(os.path.abspath(\n",
    "        os.path.join('__file__',\n",
    "                     '..',\n",
    "                     'data',\n",
    "                     'Catalogue',\n",
    "                     'Synthetic',\n",
    "                     'Cat_Anc_withBroader.tsv')),\n",
    "                   sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Return values\n",
    "The function **load_gwas_cat** returns 4 dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'return' outside function (<ipython-input-102-a8d723dc2b2b>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-102-a8d723dc2b2b>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    return Cat_Stud, Cat_Anc, Cat_Anc_byN, Cat_Full\u001b[0m\n\u001b[1;37m                                                   ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m 'return' outside function\n"
     ]
    }
   ],
   "source": [
    "#return Cat_Stud, Cat_Anc, Cat_Anc_byN, Cat_Full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze the function **load_pubmed_data()** in the *PubMed.py*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
